{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Module 1 : Unix Basics, Data Access\n",
    "\n",
    "### Class Resources\n",
    "* [Course Syllabus](http://nbviewer.ipython.org/github/bioh4x/notebooks/blob/master/home.ipynb) \n",
    "* Lead Instructor [Dr. Gene Yeo](mailto:geneyeo@ucsd.edu \"Gene Yeo\")  \n",
    "* Teaching Assistants [Olga Botvinnik](mailto:obotvinn@ucsd.edu) and [Jamison McCorrison](mailto:jmccorri@eng.ucsd.edu)  \n",
    "\n",
    "### Table of Contents\n",
    "* [**Day 1: January 5, 2016**](#Day-1-:-January-5,-2016)\n",
    "    * [Unix Basics](#Unix-Basics)\n",
    "    * [TSCC User Guide (UCSD Server Access)](#TSCC-User-Guide)\n",
    "    * [File Download and Processing Example](#File-Download-and-Processing-Example)\n",
    "    * [Homework 1](#Homework-1)\n",
    "* [**Day 2: January 7, 2016**](#Day-2:-January-7,-2016)\n",
    "    * [Git and Github](#Git-and-Github)\n",
    "    * [Downloading Data](#Downloading-Data)\n",
    "    * [Jupyter Notebook](#Jupyter-Notebook)\n",
    "    * [Python](#Python)\n",
    "    * [Addtional Resources](#Addtional-Resources)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# Day 1 : January 5, 2016\n",
    "\n",
    "* * *\n",
    "\n",
    "# Unix Basics\n",
    "\n",
    "The command line allows you to input commands, such as creating folders, deleting and copying files and extracting information from files.\n",
    "\n",
    "## Opening the command line.\n",
    "\n",
    "* **From Mac OS**\n",
    "    * Applications folder, open Utilities and launch Terminal  \n",
    "* **From Linux machine**\n",
    "    * Applications, Accessories and launch Terminal  \n",
    "* **From Windows**\n",
    "    * Determine if you have a 32 or 64 bit version of Windows. \n",
    "        * https://support.microsoft.com/en-us/kb/827218. \n",
    "        * http://cygwin.com/install.html.\n",
    "    * Run setup-x86.exe if you have 32 bit windows. If not Run setup-x86_64.exe\n",
    "    * Install Cygwin and double-click on Cygwin Terminal\n",
    "\n",
    "\n",
    "## Getting Started: Navigating your folders and files\n",
    "You start any terminal session in your \"home area\".  View your \"present working directory\"\n",
    "* `pwd`\n",
    "\n",
    "Your default home area is represented by the character alias `~` (tilde)\n",
    "* `print ~`\n",
    "\n",
    "Change directory  \n",
    "* `cd ~/Desktop`  \n",
    "\n",
    "List all the files in the present working directory using  \n",
    "* `ls`  \n",
    "* `ls .`\n",
    "    \n",
    "Arguments for unix commands  \n",
    "* `man ls`  \n",
    "    \n",
    "Creating a folder  \n",
    "* `mkdir data`  \n",
    "* `mkdir software`  \n",
    "    \n",
    "Change directory into data or software (tab complete or use Up and Down)\n",
    "* `cd da[TAB]`  \n",
    "\n",
    "Change back to the root directory from any subdirectory:\n",
    "* `cd ..`  \n",
    "\n",
    "Create an empty file\n",
    "* `touch emptyfile.txt`  \n",
    "\n",
    "Write some text in it\n",
    "* `echo \"hello world\" > emptyfile.txt`\n",
    "\n",
    "Move or rename a file\n",
    "* `mv emptyfile.txt notempty.txt`  \n",
    "\n",
    "Copy a file\n",
    "* `cp notempty.txt deleteme.txt`  \n",
    "\n",
    "Delete a file\n",
    "* `rm deleteme.txt`  \n",
    "\n",
    "Create a pointer (symlink) to a file\n",
    "* `ln -s pointer notempty.txt`  \n",
    "\n",
    "## File Manipulation: Getting some data from UCSC's Table Browser\n",
    "Go to the UCSC genome browser and the knownGene table (save as knowngene.txt)  \n",
    "* http://hgdownload.cse.ucsc.edu/downloads.html\n",
    "\n",
    "Move knownGene.txt to Desktop\n",
    "* `mv Downloads/knownGene.txt Desktop`  \n",
    "    \n",
    "(*optional*) Secure copy knownGene.txt to TSCC.\n",
    "* `scp Desktop/knownGene.txt  ewyeo@tscc-login2.sdsc.edu:.`  \n",
    "    \n",
    "What's in the file?\n",
    "* `less knownGene.txt`  \n",
    "* `cat knownGene.txt`  \n",
    "\n",
    "How many lines are in the file?\n",
    "* `wc -l knownGene.txt`  \n",
    "    \n",
    "See what's in the first n lines\n",
    "* `head -n 10 knownGene.txt`  \n",
    "    \n",
    "Check if it's indeed n lines (| command)\n",
    "* `less knownGene.txt | wc -l`\n",
    "* `wc -l knownGene.txt`  \n",
    "    \n",
    "What's in the last n lines?\n",
    "* `tail -n 10 knownGene.txt`  \n",
    "    \n",
    "Extract specific columns\n",
    "* `cut -f`  \n",
    "* `paste column1.txt column2.txt > 2columns.txt`\n",
    "\n",
    "How many genes have 3 exons?\n",
    "* `grep -c 'REGEXSEARCHTERM' target.txt`  \n",
    "    \n",
    "How many genes have 1...max # exons?\n",
    "* `sort | uniq -c`  \n",
    "    \n",
    "\n",
    "## Deleting files and file permissions\n",
    "Which user are you logged in as?\n",
    "* `whoami`\n",
    "\n",
    "What groups is that user associated with?\n",
    "* `groups`\n",
    "\n",
    "What is the ownership status of all files in my current directory?\n",
    "* `ls -lrt`\n",
    "* [Interpreting output](http://linuxcommand.org/lts0070.php)  \n",
    "\n",
    "Changing permissions\n",
    "* `chmod 775`\n",
    "\n",
    "The three digits indicate the affected user subset:\n",
    "* Front = Owner  \n",
    "* Middle = Group  \n",
    "* Rear = All Users  \n",
    "\n",
    "The value indicates visibility:  \n",
    "* 7 = May read and write  \n",
    "* 5 = May read but not write  \n",
    "* 0 = May not read or write  \n",
    "\n",
    "Changing Files Recursively\n",
    "* `chmod -R 777 Directory/`\n",
    "* `chmod -R o-rwx ~/`\n",
    "\n",
    "Changing executable nature of files\n",
    "* `chmod +x`\n",
    "\n",
    "Scratch maintenance occurs every 90 days:\n",
    "* `cd important_scratch_dir`\n",
    "* `find . | xargs touch`\n",
    "\n",
    "## Introduction to `awk`\n",
    "\n",
    "Another way to extract all lines\n",
    "* `awk -F \"\\t\" '{print;}' knownGene.txt`  \n",
    " \n",
    "What if we only wanted one column\n",
    "* `awk -F \"\\t\" '{print $8;}' knownGene.txt  | head`  \n",
    "\n",
    "What if we wanted the length of genes?\n",
    "* `awk -F \"\\t\" '{ len = $5-$4;} {print len;}' knownGene.txt | head`  \n",
    "\n",
    "Length of all genes summed?\n",
    "* `awk -F \"\\t\" '{ len = $5-$4;} {tot = tot + len;} END {print tot;}' knownGene.txt | head`  \n",
    "\n",
    "Don't process the header line (introduction to conditionals)\n",
    "* `awk -F \"\\t\" '{`\n",
    "* `if (FNR==1){`\n",
    "    *`next`\n",
    "* };\n",
    "* tot = tot + $5-$4};\n",
    "* END {print tot;}' knownGene.txt | head`\n",
    " \n",
    "What if you only want the total length of genes in chromosome 1?\n",
    "* `awk -F \"\\t\" '{`\n",
    "    * `if (FNR==1){`\n",
    "        * `next;`\n",
    "    * `};`\n",
    "    * `chr =$2;`\n",
    "    * `if (chr == \"chr1\") {`\n",
    "        * `tot = tot + $5-$4;`\n",
    "    * `}`\n",
    "* `};`\n",
    "* `END {print tot;}' knownGene.txt`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# TSCC User Guide\n",
    "\n",
    "## What is TSCC?\n",
    "\n",
    "TSCC houses the 640-core supercomputer as part of a resource sharing system which allows researchers to perform calculations and experiments when they need extra computing power.  \n",
    "  \n",
    "* TSCC user guide: http://rci.ucsd.edu/computing/index.html\n",
    "* The main contacts for questions about TSCC is the TSCC users mailing lists. The main contact for problems with TSCC is [Jim Hayes](jhayes@sdsc.edu). \n",
    "    * TSCC users: tscc-l@mailman.ucsd.edu\n",
    "    * Jim Hayes: jhayes@sdsc.edu\n",
    "\n",
    "## My First Supercomputer Login Session\n",
    "\n",
    "Your first login session will include some of the following commands, which will familiarize you with the cluster, teach you how to do some useful tasks on the queue, and help you set up a common directory structure shared by everyone in the lab.\n",
    "\n",
    "\n",
    "### 1. Log in to TSCC\n",
    "\n",
    "In your terminal, type the following (you'll need to replace username with your actual username)\n",
    "* `ssh username@tscc.sdsc.edu`\n",
    "\n",
    "### 2. Organize your home directory\n",
    "Create the base storage location for your code development (or just use your home area):\n",
    "* `mkdir code`\n",
    "* `mkdir notebooks`\n",
    "* `mkdir data`\n",
    "* `ln -s /oasis/tscc/scratch/$USER $HOME/scratch`\n",
    "\n",
    "### 3. Environment Variables and your Bash Profile\n",
    "Unix commands are written in \"BASH\".  \n",
    "\n",
    "Set a BASH environment variable\n",
    "* `export STR=\"hello world\"`  \n",
    "\n",
    "Access a variable\n",
    "* `print $STR`\n",
    "\n",
    "The most important environment variable is `$PATH`.  Folders in this path are automatically searched when looking for executable tools via auto-complete or `which`\n",
    "* `echo $PATH`\n",
    "* `which programname.sh`\n",
    "\n",
    "Customize your BASH profile by editing your `~/.bashrc` file.  This command is executed each time you log in to TSCC:\n",
    "* `source ~/.bashrc`\n",
    "\n",
    "( *optional* ) Additional details on BASH profile customization\n",
    "* https://wiki.archlinux.org/index.php/Bash\n",
    "\n",
    "### 4. Shell Scripting\n",
    "\n",
    "If you have a bunch of commands you want to run at once, you can use this script to submit them all at once. \n",
    "\n",
    "In the next example, `commands.sh` is a file has the commands you want on their own line, i.e. one command per line.\n",
    "* `java -Xms512m -Xmx512m -jar /home/yeo-lab/software/gatk/dist/Queue.jar \\`  \n",
    "`-S ~/gscripts/qscripts/do_stuff.scala --input commands.sh -run -qsub \\`  \n",
    "`-jobQueue <queue> -jobLimit <n> --ncores <n> --jobname <name> -startFromScratch`\n",
    "\n",
    "This runs a scala job that submits sub-jobs to the PBS queue under name you fill in where <name> now sits as a placeholder.\n",
    "\n",
    "### 5. Executing Tasks on a TSCC Server\n",
    "\n",
    "When you log in to TSCC, you are connected to a \"login node\".  When executing a task, you should always use an \"execution node\".  \n",
    "* More details in the TSCC user guide: http://rci.ucsd.edu/computing/index.html\n",
    "\n",
    "To submit a script that you wrote, in this case called myscript.sh, to TSCC, do:\n",
    "* `qsub -q home-yeo -l nodes=1:ppn=2 -l walltime=0:30:00 myscript.sh`\n",
    "\n",
    "To submit interactive jobs, do:\n",
    "* `qsub -I -q home-yeo -l nodes=1:ppn=2 -l walltime=0:30:00`\n",
    "\n",
    "To submit to the home-scrm queue, add -W group_list=scrm-group to your qsub command:\n",
    "* `qsub -I -l walltime=0:30:00 -q home-scrm -W group_list=scrm-group`\n",
    "\n",
    "Check the status of your jobs:\n",
    "* `qstat`\n",
    "\n",
    "Check the status of your array jobs, you need to specify ``-t`` to see the status of the individual array pieces. \n",
    "* `qstat -t`\n",
    "\n",
    "Killing jobs\n",
    "* `qdel 2006527`\n",
    "\n",
    "Kill an array job\n",
    "* `qdel 2006527[]`\n",
    "\n",
    "Kill all your jobs\n",
    "* `qdel $(qselect -u $USER)`\n",
    "\n",
    "### 6. Which queue do I submit to? (check status of queues)\n",
    "\n",
    "**Check the status of the queue** (so you know which queues to NOT submit to!)\n",
    "* `qstat -q`\n",
    "\n",
    "Example output is:  \n",
    "\n",
    "    server: tscc-mgr.local\n",
    "\n",
    "    Queue            Memory CPU Time Walltime Node  Run Que Lm  State\n",
    "    ---------------- ------ -------- -------- ----  --- --- --  -----\n",
    "    home-dkeres        --      --       --      --    2   0 --   E R\n",
    "    home-komunjer      --      --       --      --    0   0 --   E R\n",
    "    home-ong           --      --       --      --    2   0 --   E R\n",
    "    home-tg            --      --       --      --    0   0 --   E R\n",
    "    home-yeo           --      --       --      --    3   1 --   E R\n",
    "    home-visres        --      --       --      --    0   0 --   E R\n",
    "    home-mccammon      --      --       --      --   15  29 --   E R\n",
    "    home-scrm          --      --       --      --    1   0 --   E R\n",
    "    hotel              --      --    168:00:0   --  232  26 --   E R\n",
    "    home-k4zhang       --      --       --      --    0   0 --   E R\n",
    "    home-kkey          --      --       --      --    0   0 --   E R\n",
    "    home-kyang         --      --       --      --    2   1 --   E R\n",
    "    home-jsebat        --      --       --      --    1   0 --   E R\n",
    "    pdafm              --      --    72:00:00   --    1   0 --   E R\n",
    "    condo              --      --    08:00:00   --   18   6 --   E R\n",
    "    gpu-hotel          --      --    336:00:0   --    0   0 --   E R\n",
    "    glean              --      --       --      --   24  75 --   E R\n",
    "    gpu-condo          --      --    08:00:00   --   16  36 --   E R\n",
    "    home-fpaesani      --      --       --      --    4   2 --   E R\n",
    "    home-builder       --      --       --      --    0   0 --   E R\n",
    "    home               --      --       --      --    0   0 --   E R\n",
    "    home-mgilson       --      --       --      --    0   4 --   E R\n",
    "    home-eallen        --      --       --      --    0   0 --   E R\n",
    "                                                   ----- -----\n",
    "                                                     321   180\n",
    "\n",
    "So right now is not a good time to submit to the ``hotel`` queue, since it has a bunch of both running and queued jobs!\n",
    "\n",
    "**List the available Service Units** (1 SU = 1 core*hour)for a quick ego boost. Also note that our supercomputer is separated in two: yeo-group and scrm-group, but the total balance is 5.29 million SU, just enough secure us the top honors :-)\n",
    "* `gbalance | sort -nrk 3 | head`\n",
    "\n",
    "Example output is:\n",
    "\n",
    "    Id Name                 Amount  Reserved Balance CreditLimit Available\n",
    "    -- -------------------- ------- -------- ------- ----------- ---------\n",
    "    19 tideker-group        5211035    27922 5183113           0   5183113\n",
    "    82 yeo-group            3262925        0 3262925           0   3262925\n",
    "    81 scrm-group           2039328        0 2039328           0   2039328\n",
    "    14 mgilson-group         663095   208000  455095           0    455095\n",
    "    73 nanosprings-ucm       650000        0  650000           0    650000\n",
    "    17 kkey-group            635056     7104  627952           0    627952\n",
    "    16 k4zhang-group         534430        0  534430           0    534430\n",
    "    List the available TORQUE queues, for a quick boost in motivation!\n",
    "\n",
    "* `qstat -q`\n",
    "\n",
    "Example output is:\n",
    "\n",
    "    Queue            Memory CPU Time Walltime Node  Run Que Lm  State\n",
    "    ---------------- ------ -------- -------- ----  --- --- --  -----\n",
    "    home-tideker       --      --       --       16   1   0 --   E R\n",
    "    home-visres        --      --       --        1   0   0 --   E R\n",
    "    hotel              --      --    72:00:00   --   25  18 --   E R\n",
    "    home-k4zhang       --      --       --        4  21   0 --   E R\n",
    "    home-kkey          --      --       --        5   0   0 --   E R\n",
    "    pdafm              --      --    72:00:00   --    0   0 --   E R\n",
    "    condo              --      --    08:00:00   --    0   0 --   E R\n",
    "    glean              --      --       --      --    0   0 --   E R\n",
    "    home-builder       --      --       --        8   0   0 --   E R\n",
    "    home               --      --       --      --    0   0 --   E R\n",
    "    home-ewyeo         --      --       --       15   0   0 --   E R\n",
    "    home-mgilson       --      --       --        8   0   0 --   E R\n",
    "                                           ----- -----\n",
    "                                              47    18\n",
    "\n",
    "### Show available processors\n",
    "\n",
    "To show available processors\n",
    "* `showbf`\n",
    "\n",
    "Show specs of all nodes\n",
    "* `pbsnodes -a`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "#  File Download and Processing Example\n",
    "\n",
    "## In-Class Project : Gather Data From Ensembl biomart\n",
    "\n",
    "#### What is an annotation?\n",
    "An annotation is a flag denoting a special characteristic to a particular base pair or string of base pairs in a DNA sequence.  The most common type of annotation assigns particular chromosomal coordinates to a gene ID (and its associated metadata.)  Genetic sequence is most commonly stored in strings of 'A', 'C', 'G', and 'T' (fasta format).  Annotations are sometimes bundled with sequence (gbk, genbank file) or downloaded separately in memory efficient XML-based formats (gtf).\n",
    "\n",
    "#### Install the BiomaRt package (R package for simple Ensemble queries)\n",
    "[BiomaRt run guide (R)](http://www.ensembl.info/blog/2015/06/01/biomart-or-how-to-access-the-ensembl-data-from-r/)\n",
    "Download R (or Rstudio) and Install\n",
    "Start R (or RStudio).  Using R session terminal...\n",
    "Install biomaRt\n",
    "* `source(\"http://bioconductor.org/biocLite.R\")`\n",
    "* `biocLite(\"biomaRt\")`  \n",
    "      \n",
    "Test install by listing available ensembl data types, or \"marts\":\n",
    "* `library(biomaRt)`\n",
    "* `listEnsembl()`  \n",
    "      \n",
    "List filters\n",
    "* `listFilters(variation)`  \n",
    "      \n",
    "List attributes\n",
    "* `listAttributes(variation)`  \n",
    "      \n",
    "#### Gather Ensembl IDs of Interest\n",
    "Search chromosome hg19 (GRCh37) for the following attributes:\n",
    "* Ensembl gene id\n",
    "* Transcript ID\n",
    "* Aassociated gene name\n",
    "\n",
    "Example R commands:\n",
    "* `library(biomaRt)`\n",
    "* `ensembl = useEnsembl(biomart=\"ensembl\", dataset=\"hsapiens_gene_ensembl\")`\n",
    "* `genes <- getBM(attributes=c('ensembl_gene_id','gene_biotype','hgnc_symbol'), mart = ensembl)\n",
    "* `View genes`\n",
    "\n",
    "#### Convert Ensembl ids to ucsc ids\n",
    "##### Method 1:  (R)\n",
    "[Source](http://genomewiki.ucsc.edu/index.php/Gene_id_conversion)\n",
    "Biomart http://www.biomart.org is probably the best solution if your source ids are from Ensembl:\n",
    "* Click-Path:\n",
    "    * martview (top-right of screen)\n",
    "    * ensembl56 genes\n",
    "* (select your species)\n",
    "    * \"filters\"\n",
    "    * \"gene\"\n",
    "* paste your ids into \"id list limit\"\n",
    "    * \"attributes\"\n",
    "    * \"GENE\"\n",
    "* uncheck \"ensembl transcript id\"\n",
    "* uncheck \"ensembl gene id\" if you want to get rid of it\n",
    "    * \"EXTERNAL\"\n",
    "* check \"HGNC symbol\" (or \"HGNC automatic gene name\" if not human)\n",
    "\"results\"\n",
    "\n",
    "##### Method 2: (mysql)\n",
    "[Source; Pierre Lindenbaum](https://www.biostars.org/p/3570/)\n",
    "* `mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -D hg18`  \n",
    "\n",
    "For each ID ...\n",
    "* `mysql> select * from ensGene as G,ensGtp as T`\n",
    "* `where G.name=T.transcript  and T.gene=\"ENSG00000215719\" limit 1\\G`     \n",
    "\n",
    "#### Getting 3’ UTR fasta sequences with ensembl gene id as a header\n",
    "* \n",
    "(this will be useful for motif stuff later in day 2 -olga we can see which motifs from cisbp are enriched in 3’UTrs?)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# Homework 1\n",
    "\n",
    "* **TBA - UPDATE REQ'D**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# Day 2 : January 7, 2015\n",
    "    \n",
    "* * *\n",
    "\n",
    "# Git and Github\n",
    "\n",
    "* **TBA - UPDATE REQ'D**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# Downloading Data\n",
    "\n",
    "* **TBA - UPDATE REQ'D**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# Jupyter Notebook\n",
    "\n",
    "### My First Jupyter Notebook\n",
    "\n",
    "Check your python version:  \n",
    "* `python -V`\n",
    "\n",
    "Install the appropriate version of Anaconda, based on the python version:  \n",
    "* https://www.continuum.io/downloads\n",
    "        \n",
    "Install jupyter:  \n",
    "* `conda install jupyter`\n",
    "    \n",
    "Start jupyter notebook server:\n",
    "* `jupyter notebook`\n",
    "\n",
    "Connect to the jupyter notebook server:\n",
    "* http://localhost:8888/\n",
    "    \n",
    "Start a new notebook using the dropdown menu in the top right of the screen:\n",
    "![New doc image reference](newdoc.png \"New doc image reference\")\n",
    "\n",
    "### Cell Types\n",
    "\n",
    "#### Markdown and Heading:\n",
    "- Formatted text using markdown language\n",
    "    \n",
    "#### Code:\n",
    "- Input/Output dynamic processing entries\n",
    "    \n",
    "#### Raw NBConvert (Raw):\n",
    "- No input/outp;ut or markdown processing.  Unprocessed text.\n",
    "    \n",
    "**To edit the type of any cell**, select it, then use the dropdown menu at the top of the screen.\n",
    "![New doc image reference](types.png \"New doc image reference\")\n",
    "\n",
    "**To insert a new cell**, use the Insert option in the toolbar.\n",
    "![New doc image reference](insert.png \"New doc image reference\")\n",
    "\n",
    "**To edit any cell,** double-click on it.\n",
    "\n",
    "**To execute the contents of any cell (or visualize markdown language),** hit the \"execute\" button in the toolbar (play/pause symbol):\n",
    "![New doc image reference](exe.png \"New doc image reference\")\n",
    "\n",
    "### Markdown Language Basics\n",
    "\n",
    "#### Full details in the jupyter notebook user guide:\n",
    "    \n",
    "- http://jupyter.cs.brynmawr.edu/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb#4.-Using-Markdown-Cells-for-Writing\n",
    "\n",
    "#### Quick Guide to Markdown Syntax\n",
    "\n",
    "##### Headers:\n",
    "- Prepend text with \"#\" or \"##\" depending on size of desired header text (up to header size 6 = \"######\").  \n",
    "- See the 2 headers above for examples of 1 and 2 level header sizes.\n",
    "    \n",
    "##### Formatting:\n",
    "- Markdown does not automatically hard-wrp carriage returns.\n",
    "    - Insert your own break by ending the line with two spaces and then typing Return.\n",
    "- *Italics* = 1 \"*\" or \"_\"\n",
    "- **Bold** = 2 \"**\" or \"__\"\n",
    "- `monospace` text (for code) is initiated by a prepending '\\`' character\n",
    "- Use \"\\\" or a preceeding tab to remove formatting issues caused by markdown language syntax.\n",
    "    - The '\\`' monospace character also works.\n",
    "    \n",
    "    \\\\- This is not a list  \n",
    "    \\\\-- When I do this\n",
    "\n",
    "##### Quotes\n",
    "- Quotes require the use of a prepending \">\", at a count matching the quote depth.\n",
    "- \\>example\n",
    "- \\>\\>subexample\n",
    "    \n",
    "> example\n",
    ">> subexample\n",
    "    \n",
    "##### Lists\n",
    "- Lists use a \"-\", \"*\", or \"+\".  Use tabs to modify list depth.\n",
    "- Tabs reduce markdown to raw text so lists are important!  \n",
    "\n",
    "##### Links\n",
    "    Normal link example:\n",
    "    [Class Website](http://en.wikipedia.org \"Class Website\")\n",
    " \n",
    "[Class Website](http://en.wikipedia.org \"Class Website\")\n",
    "\n",
    "##### Tables\n",
    "    Minimal example:\n",
    "    SampleID|GeneID|ExpressionValue\n",
    "    -|-|-|-\n",
    "    A|ACTB|40\n",
    "    B|ACTB|9500\n",
    "    C|ACTB|0\n",
    "    \n",
    "SampleID|GeneID|ExpressionValue\n",
    "-|-|-|-\n",
    "A|ACTB|40\n",
    "B|ACTB|9500\n",
    "C|ACTB|0\n",
    "\n",
    "##### LaTeX\n",
    "- Use \"\\$\" signs to indicate LaTeX formulas:  \n",
    "    `\\${a \\choose a_1,a_2} $`\n",
    "\n",
    "${a \\choose a_1,a_2}$\n",
    "\n",
    "##### Images\n",
    "- `![Example image](http://icons.iconarchive.com/icons/icons-land/medical/256/Body-DNA-icon.png \"Example image\")`\n",
    "\n",
    "![Example image](http://icons.iconarchive.com/icons/icons-land/medical/256/Body-DNA-icon.png \"Example image\")\n",
    "\n",
    "### Intro to Dynamic Code Execution\n",
    "\n",
    "Jupyter notebook allows for test executions for over 50 programming languages within the browser.  See the full list of supported tools here:\n",
    "https://github.com/ipython/ipython/wiki/IPython-kernels-for-other-languages\n",
    "- Python is the default language\n",
    "- Use \"%%\" (magic commands) to delineate alternative languages (see below.)\n",
    "- Output is direct from the Kernel and is launched individually for each cell (ie. asycronous execution.)\n",
    "\n",
    "##### The obligatory (python) example:\n",
    "- Click on the cell below.\n",
    "- Hit the \"execute\" button (play/pause symbol) or CTRL+Enter to launch the code\n",
    "![New doc image reference](http://localhost:8888/files/Desktop/BIOM262/exe.png \"New doc image reference\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello World\n"
     ]
    }
   ],
   "source": [
    "a = \"Hello World\"\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Now let's try again using bash:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello World\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "echo \"Hello World\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Or even perl:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello World\n"
     ]
    }
   ],
   "source": [
    "%%perl\n",
    "use strict;\n",
    "use warnings;\n",
    "my $a = \"Hello World\";\n",
    "print \"$a\\n\";"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Advanced Processing Methods\n",
    "\n",
    "##### Magic commands\n",
    "* Otherwise known as \"meta commands\", these allow for code execution independent of the kernel you are using.\n",
    "* Above, the `%%bash` and `%%perl` magics were used to enter those particular kernels.\n",
    "\n",
    "##### Common magic commands\n",
    "* `?` : help command\n",
    "    * Example : `? hat` or `?? hat`\n",
    "* `!` : run as system shell\n",
    "    * Example: `! pwd` (prints present working directory)\n",
    "    * Similar to `%%bash`\n",
    "* `%magic`\n",
    "    * Lists all available magic commands \n",
    "* Example execution of magic command `pastebin` ( [Source](http://ipython.org/ipython-doc/dev/interactive/tutorial.html#magics-explained \"Source\") )\n",
    "    * `%pastebin 3 18-20 ~1/1-5`\n",
    "  \n",
    "### Advanced Processing Example with R\n",
    "\n",
    "##### Install module\n",
    "* TBA\n",
    "    \n",
    "##### Generate randomized data set based on normal distribution\n",
    "* TBA\n",
    "\n",
    "##### Render plot to Notebook\n",
    "* TBA\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# Python\n",
    "\n",
    "* **TBA - UPDATE REQ'D**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# In-Class Experiment : \n",
    "\n",
    "* X"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "# Addtional Resources\n",
    "\n",
    "* The TSCC supercomputer system primarily uses [SLURM](http://slurm.schedmd.com/) for grid processing control.\n",
    "* Jamison recommends [Oh My Zsh](http://ohmyz.sh/) as an alternative shell for your laptop.  (Not authorized for use on TSCC.)\n",
    "* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
