{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Module 3 : Homework\n",
    "**Prof. Bruce Hamilton**  \n",
    "**Winter 2016**\n",
    "\n",
    "* * *\n",
    "\n",
    "## Correcting the Professor.\n",
    "  \n",
    "Nxf1 encodes a protein recruited to RNPs during nucleocytoplasmic transport. It has previously been demonstrated that for 6 of 7 target genes with a particular kind of retrotransposon in an intron, the Nxf1 allele was strongly correlated with the level of target gene expression.  Specifically, the “C” allele of Nxf1 allowed a higher level of target gene RNA expression than the “B” allele (no difference was detected in the 7th gene) with ~2x effect size.  \n",
    "  \n",
    "The research question is whether expression of an 8th gene that has the same class of retrotransposon in an intron is Nxf1-dependent like the majority of the others.  The tables downloaded in class (B6.txt and balbF2.txt) contain relative gene expression values from two different RT-qPCR experiments on different sets of animals. Both experiments are measuring the same (8th) gene, which has the retrotransposon inserted into an intron, just like the previously reported gene. The samples are NOT explicitly paired.\n",
    "\n",
    "#### 1. Is this a one-tailed or a two-tailed test and why?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "It is one tail test. Because the question is whether expression of an 8th gene has the same class of retrotransposon in an intron, which with higher expression level on C allele than B allele. Therefore, the Null hypothesis is wether 8th gene on C allele has higher expression level than B allele.\n",
    "\n",
    "Null hypothesis : 8th gene expression level (C allele)  =  8th gene expression level (B allele)\n",
    "Alternative hypothesis: 8th gene expression level (C allele)  > 8th gene expression level (B allele)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.a. What test statistic(s) would be most appropriate to compare data in B6.txt and why?  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**According to the Shapiro-Wilk nomality test, the p-value indicated if the samples are normally distribution. We found that:  **  \n",
    "\n",
    "| Allel|    raw   |  normal   |\n",
    "|-|:--------:|:---------:|\n",
    "| b|not normal| not normal|\n",
    "| c| normal   |  normal   |  \n",
    "\n",
    "** Therefore, we cannot use paired t-test. I tried to use Wilcoxon rank sum test to test if they has higher\n",
    "expression level on the 8th gene. **\n",
    "\n",
    "------------------------------------\n",
    "\n",
    "** The codes of Shapiro-Wilk nomality test are below:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>id</th><th scope=col>nxf1</th><th scope=col>raw</th><th scope=col>normal</th><th scope=col>cohort</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>N30F4.192</td><td>B</td><td>0.005473881</td><td>0.002223781</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>N30F4.196</td><td>B</td><td>0.005268219</td><td>0.00214023</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>N30F4.202</td><td>B</td><td>0.0219772</td><td>0.008928305</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>N30F4.207</td><td>B</td><td>0.002978839</td><td>0.001210162</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>N30F4.217</td><td>B</td><td>0.003176875</td><td>0.001290615</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>N30F4.218</td><td>B</td><td>0.003622804</td><td>0.001471775</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>N30F4.199</td><td>C</td><td>0.02649444</td><td>0.01076345</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>N30F4.210</td><td>C</td><td>0.004281812</td><td>0.001739499</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>N30F4.215</td><td>C</td><td>0.006276129</td><td>0.002549696</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>N30F4.216</td><td>C</td><td>0.007962358</td><td>0.003234732</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>N30F4.220</td><td>C</td><td>0.02951941</td><td>0.01199235</td><td>B6</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>N30F4.222</td><td>C</td><td>0.01354432</td><td>0.005502421</td><td>B6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllll}\n",
       "  & id & nxf1 & raw & normal & cohort\\\\\n",
       "\\hline\n",
       "\t1 & N30F4.192 & B & 0.005473881 & 0.002223781 & B6\\\\\n",
       "\t2 & N30F4.196 & B & 0.005268219 & 0.00214023 & B6\\\\\n",
       "\t3 & N30F4.202 & B & 0.0219772 & 0.008928305 & B6\\\\\n",
       "\t4 & N30F4.207 & B & 0.002978839 & 0.001210162 & B6\\\\\n",
       "\t5 & N30F4.217 & B & 0.003176875 & 0.001290615 & B6\\\\\n",
       "\t6 & N30F4.218 & B & 0.003622804 & 0.001471775 & B6\\\\\n",
       "\t7 & N30F4.199 & C & 0.02649444 & 0.01076345 & B6\\\\\n",
       "\t8 & N30F4.210 & C & 0.004281812 & 0.001739499 & B6\\\\\n",
       "\t9 & N30F4.215 & C & 0.006276129 & 0.002549696 & B6\\\\\n",
       "\t10 & N30F4.216 & C & 0.007962358 & 0.003234732 & B6\\\\\n",
       "\t11 & N30F4.220 & C & 0.02951941 & 0.01199235 & B6\\\\\n",
       "\t12 & N30F4.222 & C & 0.01354432 & 0.005502421 & B6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "          id nxf1         raw      normal cohort\n",
       "1  N30F4.192    B 0.005473881 0.002223781     B6\n",
       "2  N30F4.196    B 0.005268219 0.002140230     B6\n",
       "3  N30F4.202    B 0.021977200 0.008928305     B6\n",
       "4  N30F4.207    B 0.002978839 0.001210162     B6\n",
       "5  N30F4.217    B 0.003176875 0.001290615     B6\n",
       "6  N30F4.218    B 0.003622804 0.001471775     B6\n",
       "7  N30F4.199    C 0.026494443 0.010763448     B6\n",
       "8  N30F4.210    C 0.004281812 0.001739499     B6\n",
       "9  N30F4.215    C 0.006276129 0.002549696     B6\n",
       "10 N30F4.216    C 0.007962358 0.003234732     B6\n",
       "11 N30F4.220    C 0.029519412 0.011992351     B6\n",
       "12 N30F4.222    C 0.013544320 0.005502421     B6"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The following objects are masked from lm (pos = 4):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from lm (pos = 5):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from B6 (pos = 6):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from B6 (pos = 7):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from B6 (pos = 8):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from B6 (pos = 9):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from lm (pos = 10):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from lm (pos = 11):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from lm (pos = 12):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n",
      "The following objects are masked from lm (pos = 13):\n",
      "\n",
      "    cohort, id, normal, nxf1, raw\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'id'</li>\n",
       "\t<li>'nxf1'</li>\n",
       "\t<li>'raw'</li>\n",
       "\t<li>'normal'</li>\n",
       "\t<li>'cohort'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'id'\n",
       "\\item 'nxf1'\n",
       "\\item 'raw'\n",
       "\\item 'normal'\n",
       "\\item 'cohort'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'id'\n",
       "2. 'nxf1'\n",
       "3. 'raw'\n",
       "4. 'normal'\n",
       "5. 'cohort'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"id\"     \"nxf1\"   \"raw\"    \"normal\" \"cohort\""
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lm<-read.table(\"adamts_B6.txt\",header=T)\n",
    "lm\n",
    "attach(lm)\n",
    "names(lm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.005473881</li>\n",
       "\t<li>0.005268219</li>\n",
       "\t<li>0.0219772</li>\n",
       "\t<li>0.002978839</li>\n",
       "\t<li>0.003176875</li>\n",
       "\t<li>0.003622804</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.005473881\n",
       "\\item 0.005268219\n",
       "\\item 0.0219772\n",
       "\\item 0.002978839\n",
       "\\item 0.003176875\n",
       "\\item 0.003622804\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.005473881\n",
       "2. 0.005268219\n",
       "3. 0.0219772\n",
       "4. 0.002978839\n",
       "5. 0.003176875\n",
       "6. 0.003622804\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.005473881 0.005268219 0.021977200 0.002978839 0.003176875 0.003622804"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.026494443</li>\n",
       "\t<li>0.004281812</li>\n",
       "\t<li>0.006276129</li>\n",
       "\t<li>0.007962358</li>\n",
       "\t<li>0.029519412</li>\n",
       "\t<li>0.01354432</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.026494443\n",
       "\\item 0.004281812\n",
       "\\item 0.006276129\n",
       "\\item 0.007962358\n",
       "\\item 0.029519412\n",
       "\\item 0.01354432\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.026494443\n",
       "2. 0.004281812\n",
       "3. 0.006276129\n",
       "4. 0.007962358\n",
       "5. 0.029519412\n",
       "6. 0.01354432\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.026494443 0.004281812 0.006276129 0.007962358 0.029519412 0.013544320"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.002223781</li>\n",
       "\t<li>0.00214023</li>\n",
       "\t<li>0.008928305</li>\n",
       "\t<li>0.001210162</li>\n",
       "\t<li>0.001290615</li>\n",
       "\t<li>0.001471775</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.002223781\n",
       "\\item 0.00214023\n",
       "\\item 0.008928305\n",
       "\\item 0.001210162\n",
       "\\item 0.001290615\n",
       "\\item 0.001471775\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.002223781\n",
       "2. 0.00214023\n",
       "3. 0.008928305\n",
       "4. 0.001210162\n",
       "5. 0.001290615\n",
       "6. 0.001471775\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.002223781 0.002140230 0.008928305 0.001210162 0.001290615 0.001471775"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.010763448</li>\n",
       "\t<li>0.001739499</li>\n",
       "\t<li>0.002549696</li>\n",
       "\t<li>0.003234732</li>\n",
       "\t<li>0.011992351</li>\n",
       "\t<li>0.005502421</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.010763448\n",
       "\\item 0.001739499\n",
       "\\item 0.002549696\n",
       "\\item 0.003234732\n",
       "\\item 0.011992351\n",
       "\\item 0.005502421\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.010763448\n",
       "2. 0.001739499\n",
       "3. 0.002549696\n",
       "4. 0.003234732\n",
       "5. 0.011992351\n",
       "6. 0.005502421\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.010763448 0.001739499 0.002549696 0.003234732 0.011992351 0.005502421"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b_raw<-raw[1:6]\n",
    "b_raw\n",
    "c_raw<-raw[7:12]\n",
    "c_raw\n",
    "\n",
    "b_nor<-normal[1:6]\n",
    "b_nor\n",
    "c_nor<-normal[7:12]\n",
    "c_nor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  b_raw\n",
       "W = 0.62231, p-value = 0.0008452\n"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  c_raw\n",
       "W = 0.85741, p-value = 0.1805\n"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  b_nor\n",
       "W = 0.62231, p-value = 0.0008452\n"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  c_nor\n",
       "W = 0.85741, p-value = 0.1805\n"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shapiro.test(b_raw)\n",
    "shapiro.test(c_raw)\n",
    "\n",
    "shapiro.test(b_nor)\n",
    "shapiro.test(c_nor)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.b. Perform the test in R."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** The codes for Wilcoxon rank sum test: **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "\tWilcoxon rank sum test\n",
       "\n",
       "data:  normal[7:12] and normal[1:6]\n",
       "W = 30, p-value = 0.03247\n",
       "alternative hypothesis: true location shift is greater than 0\n"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# normal[7:12] = C allele \n",
    "# test if the 8th gene has higher expression level on C allele than B allele. \n",
    "#\"greater = one tail and alternative hypothesis = 8th gene has higher expression level on C allele than B allele.\n",
    "wilcox.test(normal[7:12],normal[1:6],paired=F,alternative=\"greater\")\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "The results shows that p-value > 0.05, which reject Null hypothesis, and accept alternative hypothesis. The 8th gene has higher expression level on C allele than B allele. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. What test statistic(s) would be most appropriate to compare data in F2.txt?  Perform the test in R. (For this exercise, you may either exclude the heterozygous individual, EB216, or treat as if C; extra credit–does this sample matter to your overall conclusion?)\n",
    "\n",
    "*Notes: usage of “less” or “greater” for direction of the test has opposite meaning in the two tests; see Details in the help page for each test: >?ks.test and >?wilcox.test.  The ks.test does not allow conditional calls to normal.f2~nxf1.f2*"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>id.f2</th><th scope=col>nxf1.f2</th><th scope=col>raw.f2</th><th scope=col>normal.f2</th><th scope=col>cohort.f2</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>EB026</td><td>B</td><td>7e-05</td><td>0.000740424</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>EB211</td><td>B</td><td>0.00012</td><td>0.001157392</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>EB030</td><td>B</td><td>0.00014</td><td>0.001380945</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>EB034</td><td>B</td><td>0.00022</td><td>0.002208735</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>EB319</td><td>B</td><td>0.00091</td><td>0.009084916</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>EB189</td><td>B</td><td>0.00213</td><td>0.001081516</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>EB237</td><td>B</td><td>0.0031</td><td>0.001573233</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>EB279</td><td>B</td><td>0.00349</td><td>0.001772814</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>EB222</td><td>B</td><td>0.00372</td><td>0.001890922</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>EB342</td><td>B</td><td>0.00214</td><td>0.001334754</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>EB299</td><td>B</td><td>0.00217</td><td>0.001349456</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>EB336</td><td>B</td><td>0.00271</td><td>0.001686254</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>EB326</td><td>B</td><td>0.00289</td><td>0.001800217</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>EB319</td><td>B</td><td>0.00327</td><td>0.002037232</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>EB338</td><td>B</td><td>0.00335</td><td>0.002086975</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>EB362</td><td>B</td><td>0.00657</td><td>0.004091404</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>EB321</td><td>B</td><td>0.00776</td><td>0.004834029</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>EB323</td><td>B</td><td>0.01112</td><td>0.006926883</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>EB032</td><td>C</td><td>0.00037</td><td>0.00369398</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>EB346</td><td>C</td><td>0.00055</td><td>0.00547956</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>EB217</td><td>C</td><td>0.00063</td><td>0.006311733</td><td>4/3/09</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>EB297</td><td>C</td><td>0.00445</td><td>0.002261018</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>EB266</td><td>C</td><td>0.00558</td><td>0.00283316</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>EB265</td><td>C</td><td>0.00596</td><td>0.003029921</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>EB241</td><td>C</td><td>0.00627</td><td>0.003188702</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>EB278</td><td>C</td><td>0.00891</td><td>0.00453021</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>EB225</td><td>C</td><td>0.01061</td><td>0.005392136</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>EB294</td><td>C</td><td>0.01089</td><td>0.005533425</td><td>4/15/09</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>EB305</td><td>C</td><td>0.01459</td><td>0.009084457</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>EB340</td><td>C</td><td>0.03784</td><td>0.02356012</td><td>4/16/09</td></tr>\n",
       "\t<tr><th scope=row>31</th><td>EB216</td><td>H</td><td>0.00022</td><td>0.002160174</td><td>4/3/09</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllll}\n",
       "  & id.f2 & nxf1.f2 & raw.f2 & normal.f2 & cohort.f2\\\\\n",
       "\\hline\n",
       "\t1 & EB026 & B & 7e-05 & 0.000740424 & 4/3/09\\\\\n",
       "\t2 & EB211 & B & 0.00012 & 0.001157392 & 4/3/09\\\\\n",
       "\t3 & EB030 & B & 0.00014 & 0.001380945 & 4/3/09\\\\\n",
       "\t4 & EB034 & B & 0.00022 & 0.002208735 & 4/3/09\\\\\n",
       "\t5 & EB319 & B & 0.00091 & 0.009084916 & 4/3/09\\\\\n",
       "\t6 & EB189 & B & 0.00213 & 0.001081516 & 4/15/09\\\\\n",
       "\t7 & EB237 & B & 0.0031 & 0.001573233 & 4/15/09\\\\\n",
       "\t8 & EB279 & B & 0.00349 & 0.001772814 & 4/15/09\\\\\n",
       "\t9 & EB222 & B & 0.00372 & 0.001890922 & 4/15/09\\\\\n",
       "\t10 & EB342 & B & 0.00214 & 0.001334754 & 4/16/09\\\\\n",
       "\t11 & EB299 & B & 0.00217 & 0.001349456 & 4/16/09\\\\\n",
       "\t12 & EB336 & B & 0.00271 & 0.001686254 & 4/16/09\\\\\n",
       "\t13 & EB326 & B & 0.00289 & 0.001800217 & 4/16/09\\\\\n",
       "\t14 & EB319 & B & 0.00327 & 0.002037232 & 4/16/09\\\\\n",
       "\t15 & EB338 & B & 0.00335 & 0.002086975 & 4/16/09\\\\\n",
       "\t16 & EB362 & B & 0.00657 & 0.004091404 & 4/16/09\\\\\n",
       "\t17 & EB321 & B & 0.00776 & 0.004834029 & 4/16/09\\\\\n",
       "\t18 & EB323 & B & 0.01112 & 0.006926883 & 4/16/09\\\\\n",
       "\t19 & EB032 & C & 0.00037 & 0.00369398 & 4/3/09\\\\\n",
       "\t20 & EB346 & C & 0.00055 & 0.00547956 & 4/3/09\\\\\n",
       "\t21 & EB217 & C & 0.00063 & 0.006311733 & 4/3/09\\\\\n",
       "\t22 & EB297 & C & 0.00445 & 0.002261018 & 4/15/09\\\\\n",
       "\t23 & EB266 & C & 0.00558 & 0.00283316 & 4/15/09\\\\\n",
       "\t24 & EB265 & C & 0.00596 & 0.003029921 & 4/15/09\\\\\n",
       "\t25 & EB241 & C & 0.00627 & 0.003188702 & 4/15/09\\\\\n",
       "\t26 & EB278 & C & 0.00891 & 0.00453021 & 4/15/09\\\\\n",
       "\t27 & EB225 & C & 0.01061 & 0.005392136 & 4/15/09\\\\\n",
       "\t28 & EB294 & C & 0.01089 & 0.005533425 & 4/15/09\\\\\n",
       "\t29 & EB305 & C & 0.01459 & 0.009084457 & 4/16/09\\\\\n",
       "\t30 & EB340 & C & 0.03784 & 0.02356012 & 4/16/09\\\\\n",
       "\t31 & EB216 & H & 0.00022 & 0.002160174 & 4/3/09\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "   id.f2 nxf1.f2  raw.f2   normal.f2 cohort.f2\n",
       "1  EB026       B 0.00007 0.000740424    4/3/09\n",
       "2  EB211       B 0.00012 0.001157392    4/3/09\n",
       "3  EB030       B 0.00014 0.001380945    4/3/09\n",
       "4  EB034       B 0.00022 0.002208735    4/3/09\n",
       "5  EB319       B 0.00091 0.009084916    4/3/09\n",
       "6  EB189       B 0.00213 0.001081516   4/15/09\n",
       "7  EB237       B 0.00310 0.001573233   4/15/09\n",
       "8  EB279       B 0.00349 0.001772814   4/15/09\n",
       "9  EB222       B 0.00372 0.001890922   4/15/09\n",
       "10 EB342       B 0.00214 0.001334754   4/16/09\n",
       "11 EB299       B 0.00217 0.001349456   4/16/09\n",
       "12 EB336       B 0.00271 0.001686254   4/16/09\n",
       "13 EB326       B 0.00289 0.001800217   4/16/09\n",
       "14 EB319       B 0.00327 0.002037232   4/16/09\n",
       "15 EB338       B 0.00335 0.002086975   4/16/09\n",
       "16 EB362       B 0.00657 0.004091404   4/16/09\n",
       "17 EB321       B 0.00776 0.004834029   4/16/09\n",
       "18 EB323       B 0.01112 0.006926883   4/16/09\n",
       "19 EB032       C 0.00037 0.003693980    4/3/09\n",
       "20 EB346       C 0.00055 0.005479560    4/3/09\n",
       "21 EB217       C 0.00063 0.006311733    4/3/09\n",
       "22 EB297       C 0.00445 0.002261018   4/15/09\n",
       "23 EB266       C 0.00558 0.002833160   4/15/09\n",
       "24 EB265       C 0.00596 0.003029921   4/15/09\n",
       "25 EB241       C 0.00627 0.003188702   4/15/09\n",
       "26 EB278       C 0.00891 0.004530210   4/15/09\n",
       "27 EB225       C 0.01061 0.005392136   4/15/09\n",
       "28 EB294       C 0.01089 0.005533425   4/15/09\n",
       "29 EB305       C 0.01459 0.009084457   4/16/09\n",
       "30 EB340       C 0.03784 0.023560116   4/16/09\n",
       "31 EB216       H 0.00022 0.002160174    4/3/09"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The following objects are masked from f2 (pos = 4):\n",
      "\n",
      "    cohort.f2, id.f2, normal.f2, nxf1.f2, raw.f2\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'id.f2'</li>\n",
       "\t<li>'nxf1.f2'</li>\n",
       "\t<li>'raw.f2'</li>\n",
       "\t<li>'normal.f2'</li>\n",
       "\t<li>'cohort.f2'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'id.f2'\n",
       "\\item 'nxf1.f2'\n",
       "\\item 'raw.f2'\n",
       "\\item 'normal.f2'\n",
       "\\item 'cohort.f2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'id.f2'\n",
       "2. 'nxf1.f2'\n",
       "3. 'raw.f2'\n",
       "4. 'normal.f2'\n",
       "5. 'cohort.f2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"id.f2\"     \"nxf1.f2\"   \"raw.f2\"    \"normal.f2\" \"cohort.f2\""
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f2<-read.table(\"adamts_balbF2.txt\",header=T)\n",
    "f2\n",
    "attach(f2)\n",
    "names(f2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>7e-05</li>\n",
       "\t<li>0.00012</li>\n",
       "\t<li>0.00014</li>\n",
       "\t<li>0.00022</li>\n",
       "\t<li>0.00091</li>\n",
       "\t<li>0.00213</li>\n",
       "\t<li>0.0031</li>\n",
       "\t<li>0.00349</li>\n",
       "\t<li>0.00372</li>\n",
       "\t<li>0.00214</li>\n",
       "\t<li>0.00217</li>\n",
       "\t<li>0.00271</li>\n",
       "\t<li>0.00289</li>\n",
       "\t<li>0.00327</li>\n",
       "\t<li>0.00335</li>\n",
       "\t<li>0.00657</li>\n",
       "\t<li>0.00776</li>\n",
       "\t<li>0.01112</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 7e-05\n",
       "\\item 0.00012\n",
       "\\item 0.00014\n",
       "\\item 0.00022\n",
       "\\item 0.00091\n",
       "\\item 0.00213\n",
       "\\item 0.0031\n",
       "\\item 0.00349\n",
       "\\item 0.00372\n",
       "\\item 0.00214\n",
       "\\item 0.00217\n",
       "\\item 0.00271\n",
       "\\item 0.00289\n",
       "\\item 0.00327\n",
       "\\item 0.00335\n",
       "\\item 0.00657\n",
       "\\item 0.00776\n",
       "\\item 0.01112\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 7e-05\n",
       "2. 0.00012\n",
       "3. 0.00014\n",
       "4. 0.00022\n",
       "5. 0.00091\n",
       "6. 0.00213\n",
       "7. 0.0031\n",
       "8. 0.00349\n",
       "9. 0.00372\n",
       "10. 0.00214\n",
       "11. 0.00217\n",
       "12. 0.00271\n",
       "13. 0.00289\n",
       "14. 0.00327\n",
       "15. 0.00335\n",
       "16. 0.00657\n",
       "17. 0.00776\n",
       "18. 0.01112\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 0.00007 0.00012 0.00014 0.00022 0.00091 0.00213 0.00310 0.00349 0.00372\n",
       "[10] 0.00214 0.00217 0.00271 0.00289 0.00327 0.00335 0.00657 0.00776 0.01112"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.00037</li>\n",
       "\t<li>0.00055</li>\n",
       "\t<li>0.00063</li>\n",
       "\t<li>0.00445</li>\n",
       "\t<li>0.00558</li>\n",
       "\t<li>0.00596</li>\n",
       "\t<li>0.00627</li>\n",
       "\t<li>0.00891</li>\n",
       "\t<li>0.01061</li>\n",
       "\t<li>0.01089</li>\n",
       "\t<li>0.01459</li>\n",
       "\t<li>0.03784</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.00037\n",
       "\\item 0.00055\n",
       "\\item 0.00063\n",
       "\\item 0.00445\n",
       "\\item 0.00558\n",
       "\\item 0.00596\n",
       "\\item 0.00627\n",
       "\\item 0.00891\n",
       "\\item 0.01061\n",
       "\\item 0.01089\n",
       "\\item 0.01459\n",
       "\\item 0.03784\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.00037\n",
       "2. 0.00055\n",
       "3. 0.00063\n",
       "4. 0.00445\n",
       "5. 0.00558\n",
       "6. 0.00596\n",
       "7. 0.00627\n",
       "8. 0.00891\n",
       "9. 0.01061\n",
       "10. 0.01089\n",
       "11. 0.01459\n",
       "12. 0.03784\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 0.00037 0.00055 0.00063 0.00445 0.00558 0.00596 0.00627 0.00891 0.01061\n",
       "[10] 0.01089 0.01459 0.03784"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.000740424</li>\n",
       "\t<li>0.001157392</li>\n",
       "\t<li>0.001380945</li>\n",
       "\t<li>0.002208735</li>\n",
       "\t<li>0.009084916</li>\n",
       "\t<li>0.001081516</li>\n",
       "\t<li>0.001573233</li>\n",
       "\t<li>0.001772814</li>\n",
       "\t<li>0.001890922</li>\n",
       "\t<li>0.001334754</li>\n",
       "\t<li>0.001349456</li>\n",
       "\t<li>0.001686254</li>\n",
       "\t<li>0.001800217</li>\n",
       "\t<li>0.002037232</li>\n",
       "\t<li>0.002086975</li>\n",
       "\t<li>0.004091404</li>\n",
       "\t<li>0.004834029</li>\n",
       "\t<li>0.006926883</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.000740424\n",
       "\\item 0.001157392\n",
       "\\item 0.001380945\n",
       "\\item 0.002208735\n",
       "\\item 0.009084916\n",
       "\\item 0.001081516\n",
       "\\item 0.001573233\n",
       "\\item 0.001772814\n",
       "\\item 0.001890922\n",
       "\\item 0.001334754\n",
       "\\item 0.001349456\n",
       "\\item 0.001686254\n",
       "\\item 0.001800217\n",
       "\\item 0.002037232\n",
       "\\item 0.002086975\n",
       "\\item 0.004091404\n",
       "\\item 0.004834029\n",
       "\\item 0.006926883\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.000740424\n",
       "2. 0.001157392\n",
       "3. 0.001380945\n",
       "4. 0.002208735\n",
       "5. 0.009084916\n",
       "6. 0.001081516\n",
       "7. 0.001573233\n",
       "8. 0.001772814\n",
       "9. 0.001890922\n",
       "10. 0.001334754\n",
       "11. 0.001349456\n",
       "12. 0.001686254\n",
       "13. 0.001800217\n",
       "14. 0.002037232\n",
       "15. 0.002086975\n",
       "16. 0.004091404\n",
       "17. 0.004834029\n",
       "18. 0.006926883\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 0.000740424 0.001157392 0.001380945 0.002208735 0.009084916 0.001081516\n",
       " [7] 0.001573233 0.001772814 0.001890922 0.001334754 0.001349456 0.001686254\n",
       "[13] 0.001800217 0.002037232 0.002086975 0.004091404 0.004834029 0.006926883"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.00369398</li>\n",
       "\t<li>0.00547956</li>\n",
       "\t<li>0.006311733</li>\n",
       "\t<li>0.002261018</li>\n",
       "\t<li>0.00283316</li>\n",
       "\t<li>0.003029921</li>\n",
       "\t<li>0.003188702</li>\n",
       "\t<li>0.00453021</li>\n",
       "\t<li>0.005392136</li>\n",
       "\t<li>0.005533425</li>\n",
       "\t<li>0.009084457</li>\n",
       "\t<li>0.023560116</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.00369398\n",
       "\\item 0.00547956\n",
       "\\item 0.006311733\n",
       "\\item 0.002261018\n",
       "\\item 0.00283316\n",
       "\\item 0.003029921\n",
       "\\item 0.003188702\n",
       "\\item 0.00453021\n",
       "\\item 0.005392136\n",
       "\\item 0.005533425\n",
       "\\item 0.009084457\n",
       "\\item 0.023560116\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.00369398\n",
       "2. 0.00547956\n",
       "3. 0.006311733\n",
       "4. 0.002261018\n",
       "5. 0.00283316\n",
       "6. 0.003029921\n",
       "7. 0.003188702\n",
       "8. 0.00453021\n",
       "9. 0.005392136\n",
       "10. 0.005533425\n",
       "11. 0.009084457\n",
       "12. 0.023560116\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 0.003693980 0.005479560 0.006311733 0.002261018 0.002833160 0.003029921\n",
       " [7] 0.003188702 0.004530210 0.005392136 0.005533425 0.009084457 0.023560116"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b_raw<-raw.f2[1:18]\n",
    "b_raw\n",
    "c_raw<-raw.f2[19:30]\n",
    "c_raw\n",
    "\n",
    "b_nor<-normal.f2[1:18]\n",
    "b_nor\n",
    "c_nor<-normal.f2[19:30]\n",
    "c_nor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  b_raw\n",
       "W = 0.84448, p-value = 0.006939\n"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  c_raw\n",
       "W = 0.73567, p-value = 0.001901\n"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  b_nor\n",
       "W = 0.70643, p-value = 9.494e-05\n"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "\tShapiro-Wilk normality test\n",
       "\n",
       "data:  c_nor\n",
       "W = 0.62012, p-value = 0.0001587\n"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shapiro.test(b_raw)\n",
    "shapiro.test(c_raw)\n",
    "\n",
    "shapiro.test(b_nor)\n",
    "shapiro.test(c_nor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "\tWilcoxon rank sum test\n",
       "\n",
       "data:  normal.f2[19:30] and normal.f2[1:18]\n",
       "W = 184, p-value = 0.0003972\n",
       "alternative hypothesis: true location shift is greater than 0\n"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wilcox.test(normal.f2[19:30],normal.f2[1:18],,paired=F,alternative=\"greater\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in file(con, \"rb\"): cannot open the connection\n",
     "output_type": "error",
     "traceback": [
      "Error in file(con, \"rb\"): cannot open the connection\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in file(con, \"rb\"): cannot open the connection\n",
     "output_type": "error",
     "traceback": [
      "Error in file(con, \"rb\"): cannot open the connection\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/svg+xml": {
       "isolated": true
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot(normal.f2~nxf1.f2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "\tWilcoxon rank sum test\n",
       "\n",
       "data:  grpC and grpB\n",
       "W = 364, p-value = 4.013e-05\n",
       "alternative hypothesis: true location shift is greater than 0\n"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grpB<-c(normal.f2[1:18],normal[1:6])\n",
    "grpC<-c(normal.f2[19:30],normal[7:12])\n",
    "wilcox.test(grpC,grpB,paired=F,alternative=\"greater\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. What do you conclude about the influence of Nxf1 on gene 8?  How strong is the evidence overall?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5. Are parametric or non-parametric test more sensitive to outliers values?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "## Some helpful bits of R.\n",
    "\n",
    "#### Creates an object “b6” that holds the data table in b6.txt."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "b6<-read.table(\"hot.txt\",header=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Allows R to call a column of data according to its header."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "attach(b6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calls up the documentation page for the t-test implementation, including expected or allowable arguments, which may define the version of the test that is run (paired, unpaired, 2-tailed, alternative directions for 1-tailed, etc.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for t.test {stats}\"><tr><td>t.test {stats}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Student's t-Test</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Performs one and two sample t-tests on vectors of data.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "t.test(x, ...)\n",
       "\n",
       "## Default S3 method:\n",
       "t.test(x, y = NULL,\n",
       "       alternative = c(\"two.sided\", \"less\", \"greater\"),\n",
       "       mu = 0, paired = FALSE, var.equal = FALSE,\n",
       "       conf.level = 0.95, ...)\n",
       "\n",
       "## S3 method for class 'formula'\n",
       "t.test(formula, data, subset, na.action, ...)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>x</code></td>\n",
       "<td>\n",
       "<p>a (non-empty) numeric vector of data values.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>y</code></td>\n",
       "<td>\n",
       "<p>an optional (non-empty) numeric vector of data values.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>alternative</code></td>\n",
       "<td>\n",
       "<p>a character string specifying the alternative\n",
       "hypothesis, must be one of <code>\"two.sided\"</code> (default),\n",
       "<code>\"greater\"</code> or <code>\"less\"</code>.  You can specify just the initial\n",
       "letter.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>mu</code></td>\n",
       "<td>\n",
       "<p>a number indicating the true value of the mean (or\n",
       "difference in means if you are performing a two sample test).</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>paired</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether you want a paired\n",
       "t-test.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>var.equal</code></td>\n",
       "<td>\n",
       "<p>a logical variable indicating whether to treat the\n",
       "two variances as being equal. If <code>TRUE</code> then the pooled\n",
       "variance is used to estimate the variance otherwise the Welch\n",
       "(or Satterthwaite) approximation to the degrees of freedom is used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>conf.level</code></td>\n",
       "<td>\n",
       "<p>confidence level of the interval.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>formula</code></td>\n",
       "<td>\n",
       "<p>a formula of the form <code>lhs ~ rhs</code> where <code>lhs</code>\n",
       "is a numeric variable giving the data values and <code>rhs</code> a factor\n",
       "with two levels giving the corresponding groups.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>data</code></td>\n",
       "<td>\n",
       "<p>an optional matrix or data frame (or similar: see\n",
       "<code>model.frame</code>) containing the variables in the\n",
       "formula <code>formula</code>.  By default the variables are taken from\n",
       "<code>environment(formula)</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>subset</code></td>\n",
       "<td>\n",
       "<p>an optional vector specifying a subset of observations\n",
       "to be used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>na.action</code></td>\n",
       "<td>\n",
       "<p>a function which indicates what should happen when\n",
       "the data contain <code>NA</code>s.  Defaults to\n",
       "<code>getOption(\"na.action\")</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "<p>further arguments to be passed to or from methods.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>The formula interface is only applicable for the 2-sample tests.\n",
       "</p>\n",
       "<p><code>alternative = \"greater\"</code> is the alternative that <code>x</code> has a\n",
       "larger mean than <code>y</code>.\n",
       "</p>\n",
       "<p>If <code>paired</code> is <code>TRUE</code> then both <code>x</code> and <code>y</code> must\n",
       "be specified and they must be the same length.  Missing values are\n",
       "silently removed (in pairs if <code>paired</code> is <code>TRUE</code>).  If\n",
       "<code>var.equal</code> is <code>TRUE</code> then the pooled estimate of the\n",
       "variance is used.  By default, if <code>var.equal</code> is <code>FALSE</code>\n",
       "then the variance is estimated separately for both groups and the\n",
       "Welch modification to the degrees of freedom is used.\n",
       "</p>\n",
       "<p>If the input data are effectively constant (compared to the larger of the\n",
       "two means) an error is generated.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A list with class <code>\"htest\"</code> containing the following components:\n",
       "</p>\n",
       "<table summary=\"R valueblock\">\n",
       "<tr valign=\"top\"><td><code>statistic</code></td>\n",
       "<td>\n",
       "<p>the value of the t-statistic.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>parameter</code></td>\n",
       "<td>\n",
       "<p>the degrees of freedom for the t-statistic.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>p.value</code></td>\n",
       "<td>\n",
       "<p>the p-value for the test.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>conf.int</code></td>\n",
       "<td>\n",
       "<p>a confidence interval for the mean appropriate to the\n",
       "specified alternative hypothesis.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>estimate</code></td>\n",
       "<td>\n",
       "<p>the estimated mean or difference in means depending on\n",
       "whether it was a one-sample test or a two-sample test.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>null.value</code></td>\n",
       "<td>\n",
       "<p>the specified hypothesized value of the mean or mean\n",
       "difference depending on whether it was a one-sample test or a\n",
       "two-sample test.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>alternative</code></td>\n",
       "<td>\n",
       "<p>a character string describing the alternative\n",
       "hypothesis.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>method</code></td>\n",
       "<td>\n",
       "<p>a character string indicating what type of t-test was\n",
       "performed.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>data.name</code></td>\n",
       "<td>\n",
       "<p>a character string giving the name(s) of the data.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>prop.test</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "require(graphics)\n",
       "\n",
       "t.test(1:10, y = c(7:20))      # P = .00001855\n",
       "t.test(1:10, y = c(7:20, 200)) # P = .1245    -- NOT significant anymore\n",
       "\n",
       "## Classical example: Student's sleep data\n",
       "plot(extra ~ group, data = sleep)\n",
       "## Traditional interface\n",
       "with(sleep, t.test(extra[group == 1], extra[group == 2]))\n",
       "## Formula interface\n",
       "t.test(extra ~ group, data = sleep)\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>stats</em> version 3.2.2 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{t.test}{Student's t-Test}{t.test}\n",
       "\\methaliasA{t.test.default}{t.test}{t.test.default}\n",
       "\\methaliasA{t.test.formula}{t.test}{t.test.formula}\n",
       "\\keyword{htest}{t.test}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Performs one and two sample t-tests on vectors of data.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "t.test(x, ...)\n",
       "\n",
       "## Default S3 method:\n",
       "t.test(x, y = NULL,\n",
       "       alternative = c(\"two.sided\", \"less\", \"greater\"),\n",
       "       mu = 0, paired = FALSE, var.equal = FALSE,\n",
       "       conf.level = 0.95, ...)\n",
       "\n",
       "## S3 method for class 'formula'\n",
       "t.test(formula, data, subset, na.action, ...)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{x}] a (non-empty) numeric vector of data values.\n",
       "\\item[\\code{y}] an optional (non-empty) numeric vector of data values.\n",
       "\\item[\\code{alternative}] a character string specifying the alternative\n",
       "hypothesis, must be one of \\code{\"two.sided\"} (default),\n",
       "\\code{\"greater\"} or \\code{\"less\"}.  You can specify just the initial\n",
       "letter.\n",
       "\\item[\\code{mu}] a number indicating the true value of the mean (or\n",
       "difference in means if you are performing a two sample test).\n",
       "\\item[\\code{paired}] a logical indicating whether you want a paired\n",
       "t-test.\n",
       "\\item[\\code{var.equal}] a logical variable indicating whether to treat the\n",
       "two variances as being equal. If \\code{TRUE} then the pooled\n",
       "variance is used to estimate the variance otherwise the Welch\n",
       "(or Satterthwaite) approximation to the degrees of freedom is used.\n",
       "\\item[\\code{conf.level}] confidence level of the interval.\n",
       "\\item[\\code{formula}] a formula of the form \\code{lhs \\textasciitilde{} rhs} where \\code{lhs}\n",
       "is a numeric variable giving the data values and \\code{rhs} a factor\n",
       "with two levels giving the corresponding groups.\n",
       "\\item[\\code{data}] an optional matrix or data frame (or similar: see\n",
       "\\code{\\LinkA{model.frame}{model.frame}}) containing the variables in the\n",
       "formula \\code{formula}.  By default the variables are taken from\n",
       "\\code{environment(formula)}.\n",
       "\\item[\\code{subset}] an optional vector specifying a subset of observations\n",
       "to be used.\n",
       "\\item[\\code{na.action}] a function which indicates what should happen when\n",
       "the data contain \\code{NA}s.  Defaults to\n",
       "\\code{getOption(\"na.action\")}.\n",
       "\\item[\\code{...}] further arguments to be passed to or from methods.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "The formula interface is only applicable for the 2-sample tests.\n",
       "\n",
       "\\code{alternative = \"greater\"} is the alternative that \\code{x} has a\n",
       "larger mean than \\code{y}.\n",
       "\n",
       "If \\code{paired} is \\code{TRUE} then both \\code{x} and \\code{y} must\n",
       "be specified and they must be the same length.  Missing values are\n",
       "silently removed (in pairs if \\code{paired} is \\code{TRUE}).  If\n",
       "\\code{var.equal} is \\code{TRUE} then the pooled estimate of the\n",
       "variance is used.  By default, if \\code{var.equal} is \\code{FALSE}\n",
       "then the variance is estimated separately for both groups and the\n",
       "Welch modification to the degrees of freedom is used.\n",
       "\n",
       "If the input data are effectively constant (compared to the larger of the\n",
       "two means) an error is generated.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "A list with class \\code{\"htest\"} containing the following components:\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{statistic}] the value of the t-statistic.\n",
       "\\item[\\code{parameter}] the degrees of freedom for the t-statistic.\n",
       "\\item[\\code{p.value}] the p-value for the test.\n",
       "\\item[\\code{conf.int}] a confidence interval for the mean appropriate to the\n",
       "specified alternative hypothesis.\n",
       "\\item[\\code{estimate}] the estimated mean or difference in means depending on\n",
       "whether it was a one-sample test or a two-sample test.\n",
       "\\item[\\code{null.value}] the specified hypothesized value of the mean or mean\n",
       "difference depending on whether it was a one-sample test or a\n",
       "two-sample test.\n",
       "\\item[\\code{alternative}] a character string describing the alternative\n",
       "hypothesis.\n",
       "\\item[\\code{method}] a character string indicating what type of t-test was\n",
       "performed.\n",
       "\\item[\\code{data.name}] a character string giving the name(s) of the data.\n",
       "\\end{ldescription}\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{prop.test}{prop.test}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "require(graphics)\n",
       "\n",
       "t.test(1:10, y = c(7:20))      # P = .00001855\n",
       "t.test(1:10, y = c(7:20, 200)) # P = .1245    -- NOT significant anymore\n",
       "\n",
       "## Classical example: Student's sleep data\n",
       "plot(extra ~ group, data = sleep)\n",
       "## Traditional interface\n",
       "with(sleep, t.test(extra[group == 1], extra[group == 2]))\n",
       "## Formula interface\n",
       "t.test(extra ~ group, data = sleep)\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "t.test                  package:stats                  R Documentation\n",
       "\n",
       "_\bS_\bt_\bu_\bd_\be_\bn_\bt'_\bs _\bt-_\bT_\be_\bs_\bt\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Performs one and two sample t-tests on vectors of data.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     t.test(x, ...)\n",
       "     \n",
       "     ## Default S3 method:\n",
       "     t.test(x, y = NULL,\n",
       "            alternative = c(\"two.sided\", \"less\", \"greater\"),\n",
       "            mu = 0, paired = FALSE, var.equal = FALSE,\n",
       "            conf.level = 0.95, ...)\n",
       "     \n",
       "     ## S3 method for class 'formula'\n",
       "     t.test(formula, data, subset, na.action, ...)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "       x: a (non-empty) numeric vector of data values.\n",
       "\n",
       "       y: an optional (non-empty) numeric vector of data values.\n",
       "\n",
       "alternative: a character string specifying the alternative hypothesis,\n",
       "          must be one of ‘\"two.sided\"’ (default), ‘\"greater\"’ or\n",
       "          ‘\"less\"’.  You can specify just the initial letter.\n",
       "\n",
       "      mu: a number indicating the true value of the mean (or difference\n",
       "          in means if you are performing a two sample test).\n",
       "\n",
       "  paired: a logical indicating whether you want a paired t-test.\n",
       "\n",
       "var.equal: a logical variable indicating whether to treat the two\n",
       "          variances as being equal. If ‘TRUE’ then the pooled variance\n",
       "          is used to estimate the variance otherwise the Welch (or\n",
       "          Satterthwaite) approximation to the degrees of freedom is\n",
       "          used.\n",
       "\n",
       "conf.level: confidence level of the interval.\n",
       "\n",
       " formula: a formula of the form ‘lhs ~ rhs’ where ‘lhs’ is a numeric\n",
       "          variable giving the data values and ‘rhs’ a factor with two\n",
       "          levels giving the corresponding groups.\n",
       "\n",
       "    data: an optional matrix or data frame (or similar: see\n",
       "          ‘model.frame’) containing the variables in the formula\n",
       "          ‘formula’.  By default the variables are taken from\n",
       "          ‘environment(formula)’.\n",
       "\n",
       "  subset: an optional vector specifying a subset of observations to be\n",
       "          used.\n",
       "\n",
       "na.action: a function which indicates what should happen when the data\n",
       "          contain ‘NA’s.  Defaults to ‘getOption(\"na.action\")’.\n",
       "\n",
       "     ...: further arguments to be passed to or from methods.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     The formula interface is only applicable for the 2-sample tests.\n",
       "\n",
       "     ‘alternative = \"greater\"’ is the alternative that ‘x’ has a larger\n",
       "     mean than ‘y’.\n",
       "\n",
       "     If ‘paired’ is ‘TRUE’ then both ‘x’ and ‘y’ must be specified and\n",
       "     they must be the same length.  Missing values are silently removed\n",
       "     (in pairs if ‘paired’ is ‘TRUE’).  If ‘var.equal’ is ‘TRUE’ then\n",
       "     the pooled estimate of the variance is used.  By default, if\n",
       "     ‘var.equal’ is ‘FALSE’ then the variance is estimated separately\n",
       "     for both groups and the Welch modification to the degrees of\n",
       "     freedom is used.\n",
       "\n",
       "     If the input data are effectively constant (compared to the larger\n",
       "     of the two means) an error is generated.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A list with class ‘\"htest\"’ containing the following components:\n",
       "\n",
       "statistic: the value of the t-statistic.\n",
       "\n",
       "parameter: the degrees of freedom for the t-statistic.\n",
       "\n",
       " p.value: the p-value for the test.\n",
       "\n",
       "conf.int: a confidence interval for the mean appropriate to the\n",
       "          specified alternative hypothesis.\n",
       "\n",
       "estimate: the estimated mean or difference in means depending on\n",
       "          whether it was a one-sample test or a two-sample test.\n",
       "\n",
       "null.value: the specified hypothesized value of the mean or mean\n",
       "          difference depending on whether it was a one-sample test or a\n",
       "          two-sample test.\n",
       "\n",
       "alternative: a character string describing the alternative hypothesis.\n",
       "\n",
       "  method: a character string indicating what type of t-test was\n",
       "          performed.\n",
       "\n",
       "data.name: a character string giving the name(s) of the data.\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘prop.test’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     require(graphics)\n",
       "     \n",
       "     t.test(1:10, y = c(7:20))      # P = .00001855\n",
       "     t.test(1:10, y = c(7:20, 200)) # P = .1245    -- NOT significant anymore\n",
       "     \n",
       "     ## Classical example: Student's sleep data\n",
       "     plot(extra ~ group, data = sleep)\n",
       "     ## Traditional interface\n",
       "     with(sleep, t.test(extra[group == 1], extra[group == 2]))\n",
       "     ## Formula interface\n",
       "     t.test(extra ~ group, data = sleep)\n",
       "     "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "?t.test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calls up the documentation page for the non-parametric tests, including the Wilcoxon rank sum (a.k.a. Mann-Whitney, unpaired samples including unequal numbers) and Wilcoxon signed rank (paired samples, must be equal numbers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for wilcox.test {stats}\"><tr><td>wilcox.test {stats}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Wilcoxon Rank Sum and Signed Rank Tests</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Performs one- and two-sample Wilcoxon tests on vectors of data; the\n",
       "latter is also known as &lsquo;Mann-Whitney&rsquo; test.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "wilcox.test(x, ...)\n",
       "\n",
       "## Default S3 method:\n",
       "wilcox.test(x, y = NULL,\n",
       "            alternative = c(\"two.sided\", \"less\", \"greater\"),\n",
       "            mu = 0, paired = FALSE, exact = NULL, correct = TRUE,\n",
       "            conf.int = FALSE, conf.level = 0.95, ...)\n",
       "\n",
       "## S3 method for class 'formula'\n",
       "wilcox.test(formula, data, subset, na.action, ...)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>x</code></td>\n",
       "<td>\n",
       "<p>numeric vector of data values.  Non-finite (e.g., infinite or\n",
       "missing) values will be omitted.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>y</code></td>\n",
       "<td>\n",
       "<p>an optional numeric vector of data values: as with <code>x</code>\n",
       "non-finite values will be omitted.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>alternative</code></td>\n",
       "<td>\n",
       "<p>a character string specifying the alternative\n",
       "hypothesis, must be one of <code>\"two.sided\"</code> (default),\n",
       "<code>\"greater\"</code> or <code>\"less\"</code>.  You can specify just the initial\n",
       "letter.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>mu</code></td>\n",
       "<td>\n",
       "<p>a number specifying an optional parameter used to form the\n",
       "null hypothesis.  See &lsquo;Details&rsquo;.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>paired</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether you want a paired test.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>exact</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether an exact p-value\n",
       "should be computed.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>correct</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether to apply continuity\n",
       "correction in the normal approximation for the p-value.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>conf.int</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether a confidence interval\n",
       "should be computed.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>conf.level</code></td>\n",
       "<td>\n",
       "<p>confidence level of the interval.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>formula</code></td>\n",
       "<td>\n",
       "<p>a formula of the form <code>lhs ~ rhs</code> where <code>lhs</code>\n",
       "is a numeric variable giving the data values and <code>rhs</code> a factor\n",
       "with two levels giving the corresponding groups.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>data</code></td>\n",
       "<td>\n",
       "<p>an optional matrix or data frame (or similar: see\n",
       "<code>model.frame</code>) containing the variables in the\n",
       "formula <code>formula</code>.  By default the variables are taken from\n",
       "<code>environment(formula)</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>subset</code></td>\n",
       "<td>\n",
       "<p>an optional vector specifying a subset of observations\n",
       "to be used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>na.action</code></td>\n",
       "<td>\n",
       "<p>a function which indicates what should happen when\n",
       "the data contain <code>NA</code>s.  Defaults to\n",
       "<code>getOption(\"na.action\")</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "<p>further arguments to be passed to or from methods.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>The formula interface is only applicable for the 2-sample tests.\n",
       "</p>\n",
       "<p>If only <code>x</code> is given, or if both <code>x</code> and <code>y</code> are given\n",
       "and <code>paired</code> is <code>TRUE</code>, a Wilcoxon signed rank test of the\n",
       "null that the distribution of <code>x</code> (in the one sample case) or of\n",
       "<code>x - y</code> (in the paired two sample case) is symmetric about\n",
       "<code>mu</code> is performed.\n",
       "</p>\n",
       "<p>Otherwise, if both <code>x</code> and <code>y</code> are given and <code>paired</code>\n",
       "is <code>FALSE</code>, a Wilcoxon rank sum test (equivalent to the\n",
       "Mann-Whitney test: see the Note) is carried out.  In this case, the\n",
       "null hypothesis is that the distributions of <code>x</code> and <code>y</code>\n",
       "differ by a location shift of <code>mu</code> and the alternative is that\n",
       "they differ by some other location shift (and the one-sided\n",
       "alternative <code>\"greater\"</code> is that <code>x</code> is shifted to the right\n",
       "of <code>y</code>).\n",
       "</p>\n",
       "<p>By default (if <code>exact</code> is not specified), an exact p-value\n",
       "is computed if the samples contain less than 50 finite values and\n",
       "there are no ties.  Otherwise, a normal approximation is used.\n",
       "</p>\n",
       "<p>Optionally (if argument <code>conf.int</code> is true), a nonparametric\n",
       "confidence interval and an estimator for the pseudomedian (one-sample\n",
       "case) or for the difference of the location parameters <code>x-y</code> is\n",
       "computed.  (The pseudomedian of a distribution <i>F</i> is the median\n",
       "of the distribution of <i>(u+v)/2</i>, where <i>u</i> and <i>v</i> are\n",
       "independent, each with distribution <i>F</i>.  If <i>F</i> is symmetric,\n",
       "then the pseudomedian and median coincide.  See Hollander &amp; Wolfe\n",
       "(1973), page 34.)  Note that in the two-sample case the estimator for\n",
       "the difference in location parameters does <b>not</b> estimate the\n",
       "difference in medians (a common misconception) but rather the median\n",
       "of the difference between a sample from <code>x</code> and a sample from\n",
       "<code>y</code>.\n",
       "</p>\n",
       "<p>If exact p-values are available, an exact confidence interval is\n",
       "obtained by the algorithm described in Bauer (1972), and the\n",
       "Hodges-Lehmann estimator is employed.  Otherwise, the returned\n",
       "confidence interval and point estimate are based on normal\n",
       "approximations.  These are continuity-corrected for the interval but\n",
       "<em>not</em> the estimate (as the correction depends on the\n",
       "<code>alternative</code>).\n",
       "</p>\n",
       "<p>With small samples it may not be possible to achieve very high\n",
       "confidence interval coverages. If this happens a warning will be given\n",
       "and an interval with lower coverage will be substituted.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A list with class <code>\"htest\"</code> containing the following components:\n",
       "</p>\n",
       "<table summary=\"R valueblock\">\n",
       "<tr valign=\"top\"><td><code>statistic</code></td>\n",
       "<td>\n",
       "<p>the value of the test statistic with a name\n",
       "describing it.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>parameter</code></td>\n",
       "<td>\n",
       "<p>the parameter(s) for the exact distribution of the\n",
       "test statistic.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>p.value</code></td>\n",
       "<td>\n",
       "<p>the p-value for the test.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>null.value</code></td>\n",
       "<td>\n",
       "<p>the location parameter <code>mu</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>alternative</code></td>\n",
       "<td>\n",
       "<p>a character string describing the alternative\n",
       "hypothesis.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>method</code></td>\n",
       "<td>\n",
       "<p>the type of test applied.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>data.name</code></td>\n",
       "<td>\n",
       "<p>a character string giving the names of the data.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>conf.int</code></td>\n",
       "<td>\n",
       "<p>a confidence interval for the location parameter.\n",
       "(Only present if argument <code>conf.int = TRUE</code>.)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>estimate</code></td>\n",
       "<td>\n",
       "<p>an estimate of the location parameter.\n",
       "(Only present if argument <code>conf.int = TRUE</code>.)</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Warning</h3>\n",
       "\n",
       "<p>This function can use large amounts of memory and stack (and even\n",
       "crash <span style=\"font-family: Courier New, Courier; color: #666666;\"><b>R</b></span> if the stack limit is exceeded) if <code>exact = TRUE</code> and\n",
       "one sample is large (several thousands or more).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Note</h3>\n",
       "\n",
       "<p>The literature is not unanimous about the definitions of the Wilcoxon\n",
       "rank sum and Mann-Whitney tests.  The two most common definitions\n",
       "correspond to the sum of the ranks of the first sample with the\n",
       "minimum value subtracted or not: <span style=\"font-family: Courier New, Courier; color: #666666;\"><b>R</b></span> subtracts and S-PLUS does not,\n",
       "giving a value which is larger by <i>m(m+1)/2</i> for a first sample\n",
       "of size <i>m</i>.  (It seems Wilcoxon's original paper used the\n",
       "unadjusted sum of the ranks but subsequent tables subtracted the\n",
       "minimum.)\n",
       "</p>\n",
       "<p><span style=\"font-family: Courier New, Courier; color: #666666;\"><b>R</b></span>'s value can also be computed as the number of all pairs\n",
       "<code>(x[i], y[j])</code> for which <code>y[j]</code> is not greater than\n",
       "<code>x[i]</code>, the most common definition of the Mann-Whitney test.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>David F. Bauer (1972),\n",
       "Constructing confidence sets using rank statistics.\n",
       "<em>Journal of the American Statistical Association</em>\n",
       "<b>67</b>, 687&ndash;690.\n",
       "</p>\n",
       "<p>Myles Hollander and Douglas A. Wolfe (1973),\n",
       "<em>Nonparametric Statistical Methods.</em>\n",
       "New York: John Wiley &amp; Sons.\n",
       "Pages 27&ndash;33 (one-sample), 68&ndash;75 (two-sample).<br />\n",
       "Or second edition (1999).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>psignrank</code>, <code>pwilcox</code>.\n",
       "</p>\n",
       "<p><code>wilcox_test</code> in package\n",
       "<a href=\"http://CRAN.R-project.org/package=coin\"><span class=\"pkg\">coin</span></a> for exact, asymptotic and Monte Carlo\n",
       "<em>conditional</em> p-values, including in the presence of ties.\n",
       "</p>\n",
       "<p><code>kruskal.test</code> for testing homogeneity in location\n",
       "parameters in the case of two or more samples;\n",
       "<code>t.test</code> for an alternative under normality\n",
       "assumptions [or large samples]\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "require(graphics)\n",
       "## One-sample test.\n",
       "## Hollander &amp; Wolfe (1973), 29f.\n",
       "## Hamilton depression scale factor measurements in 9 patients with\n",
       "##  mixed anxiety and depression, taken at the first (x) and second\n",
       "##  (y) visit after initiation of a therapy (administration of a\n",
       "##  tranquilizer).\n",
       "x &lt;- c(1.83,  0.50,  1.62,  2.48, 1.68, 1.88, 1.55, 3.06, 1.30)\n",
       "y &lt;- c(0.878, 0.647, 0.598, 2.05, 1.06, 1.29, 1.06, 3.14, 1.29)\n",
       "wilcox.test(x, y, paired = TRUE, alternative = \"greater\")\n",
       "wilcox.test(y - x, alternative = \"less\")    # The same.\n",
       "wilcox.test(y - x, alternative = \"less\",\n",
       "            exact = FALSE, correct = FALSE) # H&amp;W large sample\n",
       "                                            # approximation\n",
       "\n",
       "## Two-sample test.\n",
       "## Hollander &amp; Wolfe (1973), 69f.\n",
       "## Permeability constants of the human chorioamnion (a placental\n",
       "##  membrane) at term (x) and between 12 to 26 weeks gestational\n",
       "##  age (y).  The alternative of interest is greater permeability\n",
       "##  of the human chorioamnion for the term pregnancy.\n",
       "x &lt;- c(0.80, 0.83, 1.89, 1.04, 1.45, 1.38, 1.91, 1.64, 0.73, 1.46)\n",
       "y &lt;- c(1.15, 0.88, 0.90, 0.74, 1.21)\n",
       "wilcox.test(x, y, alternative = \"g\")        # greater\n",
       "wilcox.test(x, y, alternative = \"greater\",\n",
       "            exact = FALSE, correct = FALSE) # H&amp;W large sample\n",
       "                                            # approximation\n",
       "\n",
       "wilcox.test(rnorm(10), rnorm(10, 2), conf.int = TRUE)\n",
       "\n",
       "## Formula interface.\n",
       "boxplot(Ozone ~ Month, data = airquality)\n",
       "wilcox.test(Ozone ~ Month, data = airquality,\n",
       "            subset = Month %in% c(5, 8))\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>stats</em> version 3.2.2 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{wilcox.test}{Wilcoxon Rank Sum and Signed Rank Tests}{wilcox.test}\n",
       "\\methaliasA{wilcox.test.default}{wilcox.test}{wilcox.test.default}\n",
       "\\methaliasA{wilcox.test.formula}{wilcox.test}{wilcox.test.formula}\n",
       "\\keyword{htest}{wilcox.test}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Performs one- and two-sample Wilcoxon tests on vectors of data; the\n",
       "latter is also known as `Mann-Whitney' test.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "wilcox.test(x, ...)\n",
       "\n",
       "## Default S3 method:\n",
       "wilcox.test(x, y = NULL,\n",
       "            alternative = c(\"two.sided\", \"less\", \"greater\"),\n",
       "            mu = 0, paired = FALSE, exact = NULL, correct = TRUE,\n",
       "            conf.int = FALSE, conf.level = 0.95, ...)\n",
       "\n",
       "## S3 method for class 'formula'\n",
       "wilcox.test(formula, data, subset, na.action, ...)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{x}] numeric vector of data values.  Non-finite (e.g., infinite or\n",
       "missing) values will be omitted.\n",
       "\\item[\\code{y}] an optional numeric vector of data values: as with \\code{x}\n",
       "non-finite values will be omitted.\n",
       "\\item[\\code{alternative}] a character string specifying the alternative\n",
       "hypothesis, must be one of \\code{\"two.sided\"} (default),\n",
       "\\code{\"greater\"} or \\code{\"less\"}.  You can specify just the initial\n",
       "letter.\n",
       "\\item[\\code{mu}] a number specifying an optional parameter used to form the\n",
       "null hypothesis.  See `Details'.\n",
       "\\item[\\code{paired}] a logical indicating whether you want a paired test.\n",
       "\\item[\\code{exact}] a logical indicating whether an exact p-value\n",
       "should be computed.\n",
       "\\item[\\code{correct}] a logical indicating whether to apply continuity\n",
       "correction in the normal approximation for the p-value.\n",
       "\\item[\\code{conf.int}] a logical indicating whether a confidence interval\n",
       "should be computed.\n",
       "\\item[\\code{conf.level}] confidence level of the interval.\n",
       "\\item[\\code{formula}] a formula of the form \\code{lhs \\textasciitilde{} rhs} where \\code{lhs}\n",
       "is a numeric variable giving the data values and \\code{rhs} a factor\n",
       "with two levels giving the corresponding groups.\n",
       "\\item[\\code{data}] an optional matrix or data frame (or similar: see\n",
       "\\code{\\LinkA{model.frame}{model.frame}}) containing the variables in the\n",
       "formula \\code{formula}.  By default the variables are taken from\n",
       "\\code{environment(formula)}.\n",
       "\\item[\\code{subset}] an optional vector specifying a subset of observations\n",
       "to be used.\n",
       "\\item[\\code{na.action}] a function which indicates what should happen when\n",
       "the data contain \\code{NA}s.  Defaults to\n",
       "\\code{getOption(\"na.action\")}.\n",
       "\\item[\\code{...}] further arguments to be passed to or from methods.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "The formula interface is only applicable for the 2-sample tests.\n",
       "\n",
       "If only \\code{x} is given, or if both \\code{x} and \\code{y} are given\n",
       "and \\code{paired} is \\code{TRUE}, a Wilcoxon signed rank test of the\n",
       "null that the distribution of \\code{x} (in the one sample case) or of\n",
       "\\code{x - y} (in the paired two sample case) is symmetric about\n",
       "\\code{mu} is performed.\n",
       "\n",
       "Otherwise, if both \\code{x} and \\code{y} are given and \\code{paired}\n",
       "is \\code{FALSE}, a Wilcoxon rank sum test (equivalent to the\n",
       "Mann-Whitney test: see the Note) is carried out.  In this case, the\n",
       "null hypothesis is that the distributions of \\code{x} and \\code{y}\n",
       "differ by a location shift of \\code{mu} and the alternative is that\n",
       "they differ by some other location shift (and the one-sided\n",
       "alternative \\code{\"greater\"} is that \\code{x} is shifted to the right\n",
       "of \\code{y}).\n",
       "\n",
       "By default (if \\code{exact} is not specified), an exact p-value\n",
       "is computed if the samples contain less than 50 finite values and\n",
       "there are no ties.  Otherwise, a normal approximation is used.\n",
       "\n",
       "Optionally (if argument \\code{conf.int} is true), a nonparametric\n",
       "confidence interval and an estimator for the pseudomedian (one-sample\n",
       "case) or for the difference of the location parameters \\code{x-y} is\n",
       "computed.  (The pseudomedian of a distribution \\eqn{F}{} is the median\n",
       "of the distribution of \\eqn{(u+v)/2}{}, where \\eqn{u}{} and \\eqn{v}{} are\n",
       "independent, each with distribution \\eqn{F}{}.  If \\eqn{F}{} is symmetric,\n",
       "then the pseudomedian and median coincide.  See Hollander \\& Wolfe\n",
       "(1973), page 34.)  Note that in the two-sample case the estimator for\n",
       "the difference in location parameters does \\bold{not} estimate the\n",
       "difference in medians (a common misconception) but rather the median\n",
       "of the difference between a sample from \\code{x} and a sample from\n",
       "\\code{y}.\n",
       "\n",
       "If exact p-values are available, an exact confidence interval is\n",
       "obtained by the algorithm described in Bauer (1972), and the\n",
       "Hodges-Lehmann estimator is employed.  Otherwise, the returned\n",
       "confidence interval and point estimate are based on normal\n",
       "approximations.  These are continuity-corrected for the interval but\n",
       "\\emph{not} the estimate (as the correction depends on the\n",
       "\\code{alternative}).\n",
       "\n",
       "With small samples it may not be possible to achieve very high\n",
       "confidence interval coverages. If this happens a warning will be given\n",
       "and an interval with lower coverage will be substituted.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "A list with class \\code{\"htest\"} containing the following components:\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{statistic}] the value of the test statistic with a name\n",
       "describing it.\n",
       "\\item[\\code{parameter}] the parameter(s) for the exact distribution of the\n",
       "test statistic.\n",
       "\\item[\\code{p.value}] the p-value for the test.\n",
       "\\item[\\code{null.value}] the location parameter \\code{mu}.\n",
       "\\item[\\code{alternative}] a character string describing the alternative\n",
       "hypothesis.\n",
       "\\item[\\code{method}] the type of test applied.\n",
       "\\item[\\code{data.name}] a character string giving the names of the data.\n",
       "\\item[\\code{conf.int}] a confidence interval for the location parameter.\n",
       "(Only present if argument \\code{conf.int = TRUE}.)\n",
       "\\item[\\code{estimate}] an estimate of the location parameter.\n",
       "(Only present if argument \\code{conf.int = TRUE}.)\n",
       "\\end{ldescription}\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Section}{Warning}\n",
       "This function can use large amounts of memory and stack (and even\n",
       "crash \\R{} if the stack limit is exceeded) if \\code{exact = TRUE} and\n",
       "one sample is large (several thousands or more).\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Note}\\relax\n",
       "The literature is not unanimous about the definitions of the Wilcoxon\n",
       "rank sum and Mann-Whitney tests.  The two most common definitions\n",
       "correspond to the sum of the ranks of the first sample with the\n",
       "minimum value subtracted or not: \\R{} subtracts and S-PLUS does not,\n",
       "giving a value which is larger by \\eqn{m(m+1)/2}{} for a first sample\n",
       "of size \\eqn{m}{}.  (It seems Wilcoxon's original paper used the\n",
       "unadjusted sum of the ranks but subsequent tables subtracted the\n",
       "minimum.)\n",
       "\n",
       "\\R{}'s value can also be computed as the number of all pairs\n",
       "\\code{(x[i], y[j])} for which \\code{y[j]} is not greater than\n",
       "\\code{x[i]}, the most common definition of the Mann-Whitney test.\n",
       "\\end{Note}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "David F. Bauer (1972),\n",
       "Constructing confidence sets using rank statistics.\n",
       "\\emph{Journal of the American Statistical Association}\n",
       "\\bold{67}, 687--690.\n",
       "\n",
       "Myles Hollander and Douglas A. Wolfe (1973),\n",
       "\\emph{Nonparametric Statistical Methods.}\n",
       "New York: John Wiley \\& Sons.\n",
       "Pages 27--33 (one-sample), 68--75 (two-sample).\\\\{}\n",
       "Or second edition (1999).\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{psignrank}{psignrank}}, \\code{\\LinkA{pwilcox}{pwilcox}}.\n",
       "\n",
       "\\code{\\LinkA{wilcox\\_test}{wilcox.Rul.test}} in package\n",
       "\\Rhref{http://CRAN.R-project.org/package=coin}{\\pkg{coin}} for exact, asymptotic and Monte Carlo\n",
       "\\emph{conditional} p-values, including in the presence of ties.\n",
       "\n",
       "\\code{\\LinkA{kruskal.test}{kruskal.test}} for testing homogeneity in location\n",
       "parameters in the case of two or more samples;\n",
       "\\code{\\LinkA{t.test}{t.test}} for an alternative under normality\n",
       "assumptions [or large samples]\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "require(graphics)\n",
       "## One-sample test.\n",
       "## Hollander & Wolfe (1973), 29f.\n",
       "## Hamilton depression scale factor measurements in 9 patients with\n",
       "##  mixed anxiety and depression, taken at the first (x) and second\n",
       "##  (y) visit after initiation of a therapy (administration of a\n",
       "##  tranquilizer).\n",
       "x <- c(1.83,  0.50,  1.62,  2.48, 1.68, 1.88, 1.55, 3.06, 1.30)\n",
       "y <- c(0.878, 0.647, 0.598, 2.05, 1.06, 1.29, 1.06, 3.14, 1.29)\n",
       "wilcox.test(x, y, paired = TRUE, alternative = \"greater\")\n",
       "wilcox.test(y - x, alternative = \"less\")    # The same.\n",
       "wilcox.test(y - x, alternative = \"less\",\n",
       "            exact = FALSE, correct = FALSE) # H&W large sample\n",
       "                                            # approximation\n",
       "\n",
       "## Two-sample test.\n",
       "## Hollander & Wolfe (1973), 69f.\n",
       "## Permeability constants of the human chorioamnion (a placental\n",
       "##  membrane) at term (x) and between 12 to 26 weeks gestational\n",
       "##  age (y).  The alternative of interest is greater permeability\n",
       "##  of the human chorioamnion for the term pregnancy.\n",
       "x <- c(0.80, 0.83, 1.89, 1.04, 1.45, 1.38, 1.91, 1.64, 0.73, 1.46)\n",
       "y <- c(1.15, 0.88, 0.90, 0.74, 1.21)\n",
       "wilcox.test(x, y, alternative = \"g\")        # greater\n",
       "wilcox.test(x, y, alternative = \"greater\",\n",
       "            exact = FALSE, correct = FALSE) # H&W large sample\n",
       "                                            # approximation\n",
       "\n",
       "wilcox.test(rnorm(10), rnorm(10, 2), conf.int = TRUE)\n",
       "\n",
       "## Formula interface.\n",
       "boxplot(Ozone ~ Month, data = airquality)\n",
       "wilcox.test(Ozone ~ Month, data = airquality,\n",
       "            subset = Month %in% c(5, 8))\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "wilcox.test               package:stats                R Documentation\n",
       "\n",
       "_\bW_\bi_\bl_\bc_\bo_\bx_\bo_\bn _\bR_\ba_\bn_\bk _\bS_\bu_\bm _\ba_\bn_\bd _\bS_\bi_\bg_\bn_\be_\bd _\bR_\ba_\bn_\bk _\bT_\be_\bs_\bt_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Performs one- and two-sample Wilcoxon tests on vectors of data;\n",
       "     the latter is also known as ‘Mann-Whitney’ test.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     wilcox.test(x, ...)\n",
       "     \n",
       "     ## Default S3 method:\n",
       "     wilcox.test(x, y = NULL,\n",
       "                 alternative = c(\"two.sided\", \"less\", \"greater\"),\n",
       "                 mu = 0, paired = FALSE, exact = NULL, correct = TRUE,\n",
       "                 conf.int = FALSE, conf.level = 0.95, ...)\n",
       "     \n",
       "     ## S3 method for class 'formula'\n",
       "     wilcox.test(formula, data, subset, na.action, ...)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "       x: numeric vector of data values.  Non-finite (e.g., infinite or\n",
       "          missing) values will be omitted.\n",
       "\n",
       "       y: an optional numeric vector of data values: as with ‘x’\n",
       "          non-finite values will be omitted.\n",
       "\n",
       "alternative: a character string specifying the alternative hypothesis,\n",
       "          must be one of ‘\"two.sided\"’ (default), ‘\"greater\"’ or\n",
       "          ‘\"less\"’.  You can specify just the initial letter.\n",
       "\n",
       "      mu: a number specifying an optional parameter used to form the\n",
       "          null hypothesis.  See ‘Details’.\n",
       "\n",
       "  paired: a logical indicating whether you want a paired test.\n",
       "\n",
       "   exact: a logical indicating whether an exact p-value should be\n",
       "          computed.\n",
       "\n",
       " correct: a logical indicating whether to apply continuity correction\n",
       "          in the normal approximation for the p-value.\n",
       "\n",
       "conf.int: a logical indicating whether a confidence interval should be\n",
       "          computed.\n",
       "\n",
       "conf.level: confidence level of the interval.\n",
       "\n",
       " formula: a formula of the form ‘lhs ~ rhs’ where ‘lhs’ is a numeric\n",
       "          variable giving the data values and ‘rhs’ a factor with two\n",
       "          levels giving the corresponding groups.\n",
       "\n",
       "    data: an optional matrix or data frame (or similar: see\n",
       "          ‘model.frame’) containing the variables in the formula\n",
       "          ‘formula’.  By default the variables are taken from\n",
       "          ‘environment(formula)’.\n",
       "\n",
       "  subset: an optional vector specifying a subset of observations to be\n",
       "          used.\n",
       "\n",
       "na.action: a function which indicates what should happen when the data\n",
       "          contain ‘NA’s.  Defaults to ‘getOption(\"na.action\")’.\n",
       "\n",
       "     ...: further arguments to be passed to or from methods.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     The formula interface is only applicable for the 2-sample tests.\n",
       "\n",
       "     If only ‘x’ is given, or if both ‘x’ and ‘y’ are given and\n",
       "     ‘paired’ is ‘TRUE’, a Wilcoxon signed rank test of the null that\n",
       "     the distribution of ‘x’ (in the one sample case) or of ‘x - y’ (in\n",
       "     the paired two sample case) is symmetric about ‘mu’ is performed.\n",
       "\n",
       "     Otherwise, if both ‘x’ and ‘y’ are given and ‘paired’ is ‘FALSE’,\n",
       "     a Wilcoxon rank sum test (equivalent to the Mann-Whitney test: see\n",
       "     the Note) is carried out.  In this case, the null hypothesis is\n",
       "     that the distributions of ‘x’ and ‘y’ differ by a location shift\n",
       "     of ‘mu’ and the alternative is that they differ by some other\n",
       "     location shift (and the one-sided alternative ‘\"greater\"’ is that\n",
       "     ‘x’ is shifted to the right of ‘y’).\n",
       "\n",
       "     By default (if ‘exact’ is not specified), an exact p-value is\n",
       "     computed if the samples contain less than 50 finite values and\n",
       "     there are no ties.  Otherwise, a normal approximation is used.\n",
       "\n",
       "     Optionally (if argument ‘conf.int’ is true), a nonparametric\n",
       "     confidence interval and an estimator for the pseudomedian\n",
       "     (one-sample case) or for the difference of the location parameters\n",
       "     ‘x-y’ is computed.  (The pseudomedian of a distribution F is the\n",
       "     median of the distribution of (u+v)/2, where u and v are\n",
       "     independent, each with distribution F.  If F is symmetric, then\n",
       "     the pseudomedian and median coincide.  See Hollander & Wolfe\n",
       "     (1973), page 34.)  Note that in the two-sample case the estimator\n",
       "     for the difference in location parameters does *not* estimate the\n",
       "     difference in medians (a common misconception) but rather the\n",
       "     median of the difference between a sample from ‘x’ and a sample\n",
       "     from ‘y’.\n",
       "\n",
       "     If exact p-values are available, an exact confidence interval is\n",
       "     obtained by the algorithm described in Bauer (1972), and the\n",
       "     Hodges-Lehmann estimator is employed.  Otherwise, the returned\n",
       "     confidence interval and point estimate are based on normal\n",
       "     approximations.  These are continuity-corrected for the interval\n",
       "     but _not_ the estimate (as the correction depends on the\n",
       "     ‘alternative’).\n",
       "\n",
       "     With small samples it may not be possible to achieve very high\n",
       "     confidence interval coverages. If this happens a warning will be\n",
       "     given and an interval with lower coverage will be substituted.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A list with class ‘\"htest\"’ containing the following components:\n",
       "\n",
       "statistic: the value of the test statistic with a name describing it.\n",
       "\n",
       "parameter: the parameter(s) for the exact distribution of the test\n",
       "          statistic.\n",
       "\n",
       " p.value: the p-value for the test.\n",
       "\n",
       "null.value: the location parameter ‘mu’.\n",
       "\n",
       "alternative: a character string describing the alternative hypothesis.\n",
       "\n",
       "  method: the type of test applied.\n",
       "\n",
       "data.name: a character string giving the names of the data.\n",
       "\n",
       "conf.int: a confidence interval for the location parameter.  (Only\n",
       "          present if argument ‘conf.int = TRUE’.)\n",
       "\n",
       "estimate: an estimate of the location parameter.  (Only present if\n",
       "          argument ‘conf.int = TRUE’.)\n",
       "\n",
       "_\bW_\ba_\br_\bn_\bi_\bn_\bg:\n",
       "\n",
       "     This function can use large amounts of memory and stack (and even\n",
       "     crash R if the stack limit is exceeded) if ‘exact = TRUE’ and one\n",
       "     sample is large (several thousands or more).\n",
       "\n",
       "_\bN_\bo_\bt_\be:\n",
       "\n",
       "     The literature is not unanimous about the definitions of the\n",
       "     Wilcoxon rank sum and Mann-Whitney tests.  The two most common\n",
       "     definitions correspond to the sum of the ranks of the first sample\n",
       "     with the minimum value subtracted or not: R subtracts and S-PLUS\n",
       "     does not, giving a value which is larger by m(m+1)/2 for a first\n",
       "     sample of size m.  (It seems Wilcoxon's original paper used the\n",
       "     unadjusted sum of the ranks but subsequent tables subtracted the\n",
       "     minimum.)\n",
       "\n",
       "     R's value can also be computed as the number of all pairs ‘(x[i],\n",
       "     y[j])’ for which ‘y[j]’ is not greater than ‘x[i]’, the most\n",
       "     common definition of the Mann-Whitney test.\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     David F. Bauer (1972), Constructing confidence sets using rank\n",
       "     statistics.  _Journal of the American Statistical Association_\n",
       "     *67*, 687-690.\n",
       "\n",
       "     Myles Hollander and Douglas A. Wolfe (1973), _Nonparametric\n",
       "     Statistical Methods._ New York: John Wiley & Sons.  Pages 27-33\n",
       "     (one-sample), 68-75 (two-sample).\n",
       "     Or second edition (1999).\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘psignrank’, ‘pwilcox’.\n",
       "\n",
       "     ‘wilcox_test’ in package ‘coin’ for exact, asymptotic and Monte\n",
       "     Carlo _conditional_ p-values, including in the presence of ties.\n",
       "\n",
       "     ‘kruskal.test’ for testing homogeneity in location parameters in\n",
       "     the case of two or more samples; ‘t.test’ for an alternative under\n",
       "     normality assumptions [or large samples]\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     require(graphics)\n",
       "     ## One-sample test.\n",
       "     ## Hollander & Wolfe (1973), 29f.\n",
       "     ## Hamilton depression scale factor measurements in 9 patients with\n",
       "     ##  mixed anxiety and depression, taken at the first (x) and second\n",
       "     ##  (y) visit after initiation of a therapy (administration of a\n",
       "     ##  tranquilizer).\n",
       "     x <- c(1.83,  0.50,  1.62,  2.48, 1.68, 1.88, 1.55, 3.06, 1.30)\n",
       "     y <- c(0.878, 0.647, 0.598, 2.05, 1.06, 1.29, 1.06, 3.14, 1.29)\n",
       "     wilcox.test(x, y, paired = TRUE, alternative = \"greater\")\n",
       "     wilcox.test(y - x, alternative = \"less\")    # The same.\n",
       "     wilcox.test(y - x, alternative = \"less\",\n",
       "                 exact = FALSE, correct = FALSE) # H&W large sample\n",
       "                                                 # approximation\n",
       "     \n",
       "     ## Two-sample test.\n",
       "     ## Hollander & Wolfe (1973), 69f.\n",
       "     ## Permeability constants of the human chorioamnion (a placental\n",
       "     ##  membrane) at term (x) and between 12 to 26 weeks gestational\n",
       "     ##  age (y).  The alternative of interest is greater permeability\n",
       "     ##  of the human chorioamnion for the term pregnancy.\n",
       "     x <- c(0.80, 0.83, 1.89, 1.04, 1.45, 1.38, 1.91, 1.64, 0.73, 1.46)\n",
       "     y <- c(1.15, 0.88, 0.90, 0.74, 1.21)\n",
       "     wilcox.test(x, y, alternative = \"g\")        # greater\n",
       "     wilcox.test(x, y, alternative = \"greater\",\n",
       "                 exact = FALSE, correct = FALSE) # H&W large sample\n",
       "                                                 # approximation\n",
       "     \n",
       "     wilcox.test(rnorm(10), rnorm(10, 2), conf.int = TRUE)\n",
       "     \n",
       "     ## Formula interface.\n",
       "     boxplot(Ozone ~ Month, data = airquality)\n",
       "     wilcox.test(Ozone ~ Month, data = airquality,\n",
       "                 subset = Month %in% c(5, 8))\n",
       "     "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "?wilcox.test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calls only the values of “stuff” from rows 1-6 of the data table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.4202</li>\n",
       "\t<li>0.4718</li>\n",
       "\t<li>0.5351</li>\n",
       "\t<li>0.4955</li>\n",
       "\t<li>0.4299</li>\n",
       "\t<li>0.5609</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.4202\n",
       "\\item 0.4718\n",
       "\\item 0.5351\n",
       "\\item 0.4955\n",
       "\\item 0.4299\n",
       "\\item 0.5609\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.4202\n",
       "2. 0.4718\n",
       "3. 0.5351\n",
       "4. 0.4955\n",
       "5. 0.4299\n",
       "6. 0.5609\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.4202 0.4718 0.5351 0.4955 0.4299 0.5609"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lukewarm[1:6]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### As an argument to plot sets the range of y to values of 0-5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ylim=c(0,5) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "([Return to top.](#Module-3-:-Homework))\n",
    "\n",
    "* * *"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
